// @temple-digital-group/temple-canton-js/dist/api/index.js
import axios from "axios";
import config, { initializeConfig, setWalletAdapter } from "../../src/config/index.js";
import { getAuthHeader, setUserId, } from "./tokenStore.js";
export { getUserId } from "./tokenStore.js";
export { setWalletAdapter } from "../../src/config/index.js";
// ── Initialization ──
/**
 * Initialize the Temple SDK.
 * API_KEY is required. Throws if not provided.
 */
export function initialize(cfg) {
    if (!cfg.API_KEY) {
        throw new Error("initialize: API_KEY is required.");
    }
    initializeConfig(cfg);
    if (cfg.WALLET_ADAPTER) {
        setWalletAdapter(cfg.WALLET_ADAPTER);
    }
    const userIdValue = cfg.USER_ID;
    if (userIdValue) {
        setUserId(userIdValue);
    }
}
// ── Internal helpers ──
function handleApiError(error, context) {
    const err = error;
    const status = err.response?.status ?? null;
    const message = err.response?.data?.message ??
        err.response?.data?.error ??
        err.message ??
        "Unknown error";
    const code = err.response?.data?.code ?? null;
    console.error(`[Temple API] ${context}: ${message}${status ? ` (HTTP ${status})` : ""}`);
    return { error: true, status, code, message };
}
function ensureAuthenticated() {
    if (config.API_KEY)
        return null;
    return { error: true, status: 401, code: "NOT_AUTHENTICATED", message: "API key required. Pass API_KEY in initialize()." };
}
async function authenticatedGet(path, params) {
    const authError = ensureAuthenticated();
    if (authError)
        return authError;
    const headers = {
        ...getAuthHeader(),
        "Content-Type": "application/json",
    };
    // Filter out undefined params
    const cleanParams = {};
    if (params) {
        for (const [key, value] of Object.entries(params)) {
            if (value !== undefined)
                cleanParams[key] = value;
        }
    }
    try {
        const response = await axios.get(`${config.API_BASE_URL}${path}`, {
            headers,
            params: Object.keys(cleanParams).length > 0 ? cleanParams : undefined,
        });
        return response.data;
    }
    catch (error) {
        return handleApiError(error, `GET ${path}`);
    }
}
async function authenticatedPost(path, body) {
    const authError = ensureAuthenticated();
    if (authError)
        return authError;
    const headers = {
        ...getAuthHeader(),
        "Content-Type": "application/json",
    };
    try {
        const response = await axios.post(`${config.API_BASE_URL}${path}`, body ?? {}, { headers });
        return response.data;
    }
    catch (error) {
        return handleApiError(error, `POST ${path}`);
    }
}
// ── Helpers ──
/** Normalize CC → Amulet in symbol strings (e.g. "CC/SBC" → "Amulet/SBC", "CC" → "Amulet") */
function normalizeSymbol(symbol) {
    return symbol.replace(/\bCC\b/g, "Amulet");
}
// ── Market Data ──
/**
 * Get ticker data for one or all trading pairs.
 * @param symbol - Optional symbol filter (e.g., "Amulet/USDCx")
 */
export async function getTicker(symbol) {
    return authenticatedGet("/api/v1/market/ticker", { symbol: symbol ? normalizeSymbol(symbol) : undefined });
}
/**
 * Get the order book for a trading pair.
 * @param symbol - Trading pair symbol (required)
 * @param options - Optional levels and precision parameters
 */
export async function getOrderBook(symbol, options = {}) {
    if (!symbol) {
        return { error: true, status: null, code: "INVALID_PARAMS", message: "Symbol is required." };
    }
    return authenticatedGet("/api/v1/market/orderbook", {
        symbol: normalizeSymbol(symbol),
        levels: options.levels,
        precision: options.precision,
    });
}
/**
 * Get symbol configuration (paused status, decimals, minimum quantity).
 * @param symbol - Trading pair symbol (required)
 */
export async function getSymbolConfig(symbol) {
    if (!symbol) {
        return { error: true, status: null, code: "INVALID_PARAMS", message: "Symbol is required." };
    }
    return authenticatedGet("/api/v1/trading/symbol-config", { symbol: normalizeSymbol(symbol) });
}
/**
 * Get open interest for a trading pair.
 * @param symbol - Trading pair symbol (required)
 */
export async function getOpenInterest(symbol) {
    if (!symbol) {
        return { error: true, status: null, code: "INVALID_PARAMS", message: "Symbol is required." };
    }
    return authenticatedGet("/api/v1/market/open-interest", { symbol: normalizeSymbol(symbol) });
}
/**
 * Get recent trades for a trading pair.
 * @param symbol - Trading pair symbol (required)
 * @param options - Optional limit (max 500)
 */
export async function getRecentTrades(symbol, options = {}) {
    if (!symbol) {
        return { error: true, status: null, code: "INVALID_PARAMS", message: "Symbol is required." };
    }
    return authenticatedGet("/api/v1/market/trades", {
        symbol: normalizeSymbol(symbol),
        limit: options.limit,
    });
}
// ── Orders ──
/**
 * Get active orders, optionally filtered by symbol.
 * @param options - Optional symbol and limit filters
 */
export async function getActiveOrders(options = {}) {
    return authenticatedGet("/api/trading/orders/active", {
        symbol: options.symbol ? normalizeSymbol(options.symbol) : undefined,
        limit: options.limit,
    });
}
/**
 * Cancel a specific order by ID.
 * @param orderId - The order ID to cancel
 */
export async function cancelOrder(orderId) {
    if (!orderId) {
        return { error: true, status: null, code: "INVALID_PARAMS", message: "Order ID is required." };
    }
    return authenticatedPost(`/api/trading/orders/${encodeURIComponent(orderId)}/cancel`);
}
/**
 * Cancel all active orders, optionally filtered by symbol.
 * @param options - Optional symbol filter
 */
export async function cancelAllOrders(options = {}) {
    return authenticatedPost("/api/trading/orders/cancel-all", options.symbol ? { symbol: normalizeSymbol(options.symbol) } : undefined);
}
// ── Disclosures ──
/**
 * Get amulet disclosure data for a party from the Temple REST API.
 * Returns protocol disclosures (amulet rules, open mining rounds, external party amulet rules).
 * @param partyId - Canton party ID
 */
export async function getDisclosures(partyId) {
    if (!partyId) {
        return { error: true, status: null, code: "INVALID_PARAMS", message: "Party ID is required." };
    }
    return authenticatedGet("/api/amulet/disclosures", { partyId });
}
// ── Delegation ──
/**
 * Get the user's trading delegation status.
 */
export async function getDelegation() {
    return authenticatedGet("/api/trading/delegation");
}
// ── Order Requests ──
/**
 * Create an order request (limit buy/sell) on the Temple trading backend.
 * @param opts - Order parameters: symbol, side, quantity, price, order_type, expires_at
 */
export async function createOrderRequest(opts) {
    const { symbol: rawSymbol, side, quantity, price, order_type = "limit", expires_at } = opts || {};
    if (!rawSymbol || !side || quantity == null || price == null) {
        return { error: true, status: null, code: "INVALID_PARAMS", message: "symbol, side, quantity, and price are required." };
    }
    const symbol = normalizeSymbol(rawSymbol);
    const body = { symbol, side, quantity: Number(quantity), price: Number(price), order_type };
    if (expires_at)
        body.expires_at = expires_at;
    return authenticatedPost("/api/trading/orders", body);
}
// ── Withdrawals ──
/**
 * Create a withdrawal request.
 * @param assetId - Instrument symbol (e.g. "USDCx", "Amulet")
 * @param amount - Amount to withdraw
 */
export async function createWithdrawalRequest(assetId, amount) {
    if (!assetId || amount == null) {
        return { error: true, status: null, code: "INVALID_PARAMS", message: "asset_id and amount are required." };
    }
    return authenticatedPost("/api/trading/withdrawals", { asset_id: normalizeSymbol(assetId), amount: String(amount) });
}
/**
 * Get the status of a withdrawal request.
 * @param requestId - The request ID returned by createWithdrawalRequest
 */
export async function getWithdrawalRequestStatus(requestId) {
    if (requestId == null) {
        return { error: true, status: null, code: "INVALID_PARAMS", message: "requestId is required." };
    }
    return authenticatedGet(`/api/trading/withdrawals/${encodeURIComponent(requestId)}`);
}
/**
 * Get the user's trading balances.
 *
 * GET /api/trading/balances
 */
export async function getTradingBalance() {
    return authenticatedGet("/api/trading/balances");
}
