// @temple-digital-group/temple-canton-js/src/api/index.ts
import axios, { type AxiosRequestConfig } from "axios";
import config, { initializeConfig, setWalletAdapter } from "../../src/config/index.js";
import {
	getAuthHeader,
	setUserId,
} from "./tokenStore.js";
import type {
	ApiError,
	Ticker,
	OrderBook,
	OrderBookOptions,
	SymbolConfig,
	OpenInterest,
	Trade,
	RecentTradesOptions,
	ActiveOrder,
	ActiveOrdersOptions,
	CancelOrderResponse,
	CancelAllOrdersOptions,
	CancelAllOrdersResponse,
	CreateOrderRequestOpts,
	CreateOrderRequestResponse,
	DelegationResponse,
	WithdrawalResponse,
	WithdrawalStatusResponse,
	TradingBalanceResponse,
	AmuletDisclosure,
	DisclosuresResponse,
} from "./types.js";

// Re-export types and token utilities consumers may need
export type {
	ApiError,
	Ticker,
	OrderBook,
	OrderBookOptions,
	SymbolConfig,
	OpenInterest,
	Trade,
	RecentTradesOptions,
	ActiveOrder,
	ActiveOrdersOptions,
	CancelOrderResponse,
	CancelAllOrdersOptions,
	CancelAllOrdersResponse,
	CreateOrderRequestResponse,
	DelegationResponse,
	WithdrawalResponse,
	WithdrawalStatusResponse,
	TradingBalanceResponse,
	AmuletDisclosure,
	DisclosuresResponse,
};
export { getUserId } from "./tokenStore.js";
export { setWalletAdapter } from "../../src/config/index.js";

// ── Initialization ──

/**
 * Initialize the Temple SDK.
 * API_KEY is required. Throws if not provided.
 */
export function initialize(cfg: Record<string, unknown>): void {
	if (!cfg.API_KEY) {
		throw new Error("initialize: API_KEY is required.");
	}

	initializeConfig(cfg);

	if (cfg.WALLET_ADAPTER) {
		setWalletAdapter(cfg.WALLET_ADAPTER);
	}

	const userIdValue = cfg.USER_ID as string | undefined;
	if (userIdValue) {
		setUserId(userIdValue);
	}
}

// ── Internal helpers ──

function handleApiError(error: unknown, context: string): ApiError {
	const err = error as { response?: { status?: number; data?: { message?: string; error?: string; code?: string } }; message?: string };
	const status = err.response?.status ?? null;
	const message =
		err.response?.data?.message ??
		err.response?.data?.error ??
		err.message ??
		"Unknown error";
	const code = err.response?.data?.code ?? null;

	console.error(`[Temple API] ${context}: ${message}${status ? ` (HTTP ${status})` : ""}`);

	return { error: true, status, code, message };
}

function ensureAuthenticated(): ApiError | null {
	if (config.API_KEY) return null;
	return { error: true, status: 401, code: "NOT_AUTHENTICATED", message: "API key required. Pass API_KEY in initialize()." };
}

async function authenticatedGet<T>(path: string, params?: Record<string, string | number | undefined>): Promise<T | ApiError> {
	const authError = ensureAuthenticated();
	if (authError) return authError;

	const headers = {
		...getAuthHeader(),
		"Content-Type": "application/json",
	};

	// Filter out undefined params
	const cleanParams: Record<string, string | number> = {};
	if (params) {
		for (const [key, value] of Object.entries(params)) {
			if (value !== undefined) cleanParams[key] = value;
		}
	}

	try {
		const response = await axios.get<T>(`${config.API_BASE_URL}${path}`, {
			headers,
			params: Object.keys(cleanParams).length > 0 ? cleanParams : undefined,
		} as AxiosRequestConfig);
		return response.data;
	} catch (error) {
		return handleApiError(error, `GET ${path}`);
	}
}

async function authenticatedPost<T>(path: string, body?: Record<string, unknown>): Promise<T | ApiError> {
	const authError = ensureAuthenticated();
	if (authError) return authError;

	const headers = {
		...getAuthHeader(),
		"Content-Type": "application/json",
	};

	try {
		const response = await axios.post<T>(`${config.API_BASE_URL}${path}`, body ?? {}, { headers } as AxiosRequestConfig);
		return response.data;
	} catch (error) {
		return handleApiError(error, `POST ${path}`);
	}
}

// ── Helpers ──

/** Normalize CC → Amulet in symbol strings (e.g. "CC/SBC" → "Amulet/SBC", "CC" → "Amulet") */
function normalizeSymbol(symbol: string): string {
	return symbol.replace(/\bCC\b/g, "Amulet");
}

// ── Market Data ──

/**
 * Get ticker data for one or all trading pairs.
 * @param symbol - Optional symbol filter (e.g., "Amulet/USDCx")
 */
export async function getTicker(symbol?: string): Promise<Ticker | Ticker[] | ApiError> {
	return authenticatedGet("/api/v1/market/ticker", { symbol: symbol ? normalizeSymbol(symbol) : undefined });
}

/**
 * Get the order book for a trading pair.
 * @param symbol - Trading pair symbol (required)
 * @param options - Optional levels and precision parameters
 */
export async function getOrderBook(symbol: string, options: OrderBookOptions = {}): Promise<OrderBook | ApiError> {
	if (!symbol) {
		return { error: true, status: null, code: "INVALID_PARAMS", message: "Symbol is required." };
	}
	return authenticatedGet("/api/v1/market/orderbook", {
		symbol: normalizeSymbol(symbol),
		levels: options.levels,
		precision: options.precision,
	});
}

/**
 * Get symbol configuration (paused status, decimals, minimum quantity).
 * @param symbol - Trading pair symbol (required)
 */
export async function getSymbolConfig(symbol: string): Promise<SymbolConfig | ApiError> {
	if (!symbol) {
		return { error: true, status: null, code: "INVALID_PARAMS", message: "Symbol is required." };
	}
	return authenticatedGet("/api/v1/trading/symbol-config", { symbol: normalizeSymbol(symbol) });
}

/**
 * Get open interest for a trading pair.
 * @param symbol - Trading pair symbol (required)
 */
export async function getOpenInterest(symbol: string): Promise<OpenInterest | ApiError> {
	if (!symbol) {
		return { error: true, status: null, code: "INVALID_PARAMS", message: "Symbol is required." };
	}
	return authenticatedGet("/api/v1/market/open-interest", { symbol: normalizeSymbol(symbol) });
}

/**
 * Get recent trades for a trading pair.
 * @param symbol - Trading pair symbol (required)
 * @param options - Optional limit (max 500)
 */
export async function getRecentTrades(symbol: string, options: RecentTradesOptions = {}): Promise<Trade[] | ApiError> {
	if (!symbol) {
		return { error: true, status: null, code: "INVALID_PARAMS", message: "Symbol is required." };
	}
	return authenticatedGet("/api/v1/market/trades", {
		symbol: normalizeSymbol(symbol),
		limit: options.limit,
	});
}

// ── Orders ──

/**
 * Get active orders, optionally filtered by symbol.
 * @param options - Optional symbol and limit filters
 */
export async function getActiveOrders(options: ActiveOrdersOptions = {}): Promise<ActiveOrder[] | ApiError> {
	return authenticatedGet("/api/trading/orders/active", {
		symbol: options.symbol ? normalizeSymbol(options.symbol) : undefined,
		limit: options.limit,
	});
}

/**
 * Cancel a specific order by ID.
 * @param orderId - The order ID to cancel
 */
export async function cancelOrder(orderId: string): Promise<CancelOrderResponse | ApiError> {
	if (!orderId) {
		return { error: true, status: null, code: "INVALID_PARAMS", message: "Order ID is required." };
	}
	return authenticatedPost(`/api/trading/orders/${encodeURIComponent(orderId)}/cancel`);
}

/**
 * Cancel all active orders, optionally filtered by symbol.
 * @param options - Optional symbol filter
 */
export async function cancelAllOrders(options: CancelAllOrdersOptions = {}): Promise<CancelAllOrdersResponse | ApiError> {
	return authenticatedPost("/api/trading/orders/cancel-all", options.symbol ? { symbol: normalizeSymbol(options.symbol) } : undefined);
}

// ── Disclosures ──

/**
 * Get amulet disclosure data for a party from the Temple REST API.
 * Returns protocol disclosures (amulet rules, open mining rounds, external party amulet rules).
 * @param partyId - Canton party ID
 */
export async function getDisclosures(partyId: string): Promise<DisclosuresResponse | ApiError> {
	if (!partyId) {
		return { error: true, status: null, code: "INVALID_PARAMS", message: "Party ID is required." };
	}
	return authenticatedGet("/api/amulet/disclosures", { partyId });
}

// ── Delegation ──

/**
 * Get the user's trading delegation status.
 */
export async function getDelegation(): Promise<DelegationResponse | ApiError> {
	return authenticatedGet("/api/trading/delegation");
}

// ── Order Requests ──

/**
 * Create an order request (limit buy/sell) on the Temple trading backend.
 * @param opts - Order parameters: symbol, side, quantity, price, order_type, expires_at
 */
export async function createOrderRequest(opts: CreateOrderRequestOpts): Promise<CreateOrderRequestResponse | ApiError> {
	const { symbol: rawSymbol, side, quantity, price, order_type = "limit", expires_at } = opts || {};
	if (!rawSymbol || !side || quantity == null || price == null) {
		return { error: true, status: null, code: "INVALID_PARAMS", message: "symbol, side, quantity, and price are required." };
	}
	const symbol = normalizeSymbol(rawSymbol);
	const body: Record<string, unknown> = { symbol, side, quantity: Number(quantity), price: Number(price), order_type };
	if (expires_at) body.expires_at = expires_at;
	return authenticatedPost("/api/trading/orders", body);
}

// ── Withdrawals ──

/**
 * Create a withdrawal request.
 * @param assetId - Instrument symbol (e.g. "USDCx", "Amulet")
 * @param amount - Amount to withdraw
 */
export async function createWithdrawalRequest(assetId: string, amount: string | number): Promise<WithdrawalResponse | ApiError> {
	if (!assetId || amount == null) {
		return { error: true, status: null, code: "INVALID_PARAMS", message: "asset_id and amount are required." };
	}
	return authenticatedPost("/api/trading/withdrawals", { asset_id: normalizeSymbol(assetId), amount: String(amount) });
}

/**
 * Get the status of a withdrawal request.
 * @param requestId - The request ID returned by createWithdrawalRequest
 */
export async function getWithdrawalRequestStatus(requestId: string | number): Promise<WithdrawalStatusResponse | ApiError> {
	if (requestId == null) {
		return { error: true, status: null, code: "INVALID_PARAMS", message: "requestId is required." };
	}
	return authenticatedGet(`/api/trading/withdrawals/${encodeURIComponent(requestId)}`);
}

/**
 * Get the user's trading balances.
 *
 * GET /api/trading/balances
 */
export async function getTradingBalance(): Promise<TradingBalanceResponse | ApiError> {
	return authenticatedGet("/api/trading/balances");
}
