// @temple-digital-group/temple-canton-js/src/canton/index.js
import config, { NETWORK_LOCALHOST } from "../config/index.js";
import axios from "axios";
import { getDisclosures, getDelegation } from "../../dist/api/index.js";
import { getUserId } from "../../dist/api/tokenStore.js";
import { getAdapterPartyId, getWalletAdapter, submitCommand } from "./walletAdapter.js";

// Import shared helpers from compiled TypeScript module
import {
	randomUUID,
	shouldUseLedgerForMetadata,
	normalizeContractId,
	normalizeContractReferences,
	resolveProvider,
	resolveInstrumentDefinition,
	getInstrumentRegistrar,
	unwrapActiveContracts,
	findDisclosedContract,
	mapActiveContractMetadata,
	toContractMetadata,
	dedupeDisclosedContracts,
	buildHeaders,
	getLedgerEnd,
	DEFAULT_AMULET_CONTEXT_KEYS,
	DEFAULT_UTILITY_CONTEXT_KEYS,
} from "../../dist/canton/helpers.js";

// Re-export helpers that are part of the public API
export { resolveInstrumentDefinition, getInstrumentRegistrar };

// Import JSON schemas
import createOrderProposalAmuletSchema from "./request_schemas/create_order_proposal_amulet.json" with { type: "json" };
import createOrderProposalUtilitySchema from "./request_schemas/create_order_proposal_utility.json" with { type: "json" };
import getAmuletHoldingsSchema from "./request_schemas/get_amulet_holdings.json" with { type: "json" };
import getLockedAmuletHoldingsSchema from "./request_schemas/get_locked_amulet_holdings.json" with { type: "json" };
import getUtilityHoldingsSchema from "./request_schemas/get_utility_holdings.json" with { type: "json" };
import getUtilityInstrumentConfigurationsSchema from "./request_schemas/get_instrument_configurations.json" with { type: "json" };
import getUtilityAllocationFactorySchema from "./request_schemas/get_allocation_factory.json" with { type: "json" };
import getUtilitySenderCredentialsSchema from "./request_schemas/get_sender_credentials.json" with { type: "json" };
import getOrderProposalsSchema from "./request_schemas/get_order_proposals.json" with { type: "json" };
import getOrdersSchema from "./request_schemas/get_orders.json" with { type: "json" };
import getTransferFactorySchema from "./request_schemas/get_transfer_factory.json" with { type: "json" };
import executeTransferFactorySchema from "./request_schemas/execute_transfer_factory.json" with { type: "json" };
import unlockAmuletSchema from "./request_schemas/unlock_amulet.json" with { type: "json" };
import createUtilityCredentialSchema from "./request_schemas/create_utility_credential.json" with { type: "json" };

// Import instrument catalog
import { RegistrarInternalScheme, instrumentCatalog, instrumentIdToSymbol, supportedTradingPairs, supportedSymbols } from "./instrumentCatalog.js";


/**
 * Get the list of supported trading pairs.
 * @returns {Array<{baseAsset: string, quoteAsset: string}>} Array of trading pairs
 */
export function getSupportedTradingPairs() {
	return [...supportedTradingPairs];
}

/**
 * Get the instrument catalog with all supported instruments.
 * @returns {Object} Map of asset IDs to instrument definitions
 */
export function getInstrumentCatalog() {
	return { ...instrumentCatalog };
}


function resolveContextKeyValue(...candidates) {
	for (const value of candidates) {
		if (typeof value === "string" && value.trim().length > 0) {
			return value.trim();
		}
	}
	return null;
}

function buildAmuletContextKeys(source = {}) {
	const amuletRules = resolveContextKeyValue(source.amuletRules, source.amulet_rules, DEFAULT_AMULET_CONTEXT_KEYS.amuletRules);
	const featuredAppRight = resolveContextKeyValue(source.featuredAppRight, source.featured_app_right, DEFAULT_AMULET_CONTEXT_KEYS.featuredAppRight);
	const openRound = resolveContextKeyValue(source.openRound, source.open_round, DEFAULT_AMULET_CONTEXT_KEYS.openRound);
	const expireLock = resolveContextKeyValue(source.expireLock, source.expire_lock, DEFAULT_AMULET_CONTEXT_KEYS.expireLock);

	return {
		amuletRules,
		amulet_rules: amuletRules,
		featuredAppRight,
		featured_app_right: featuredAppRight,
		openRound,
		open_round: openRound,
		expireLock,
		expire_lock: expireLock,
	};
}

function matchesDso(payload, dso) {
	if (!payload || !dso) {
		return false;
	}
	const target = dso.trim();
	if (!target) {
		return false;
	}
	const candidates = [payload.dso, payload.provider, payload.admin, payload.dsoParty, payload.rulesProvider, payload.dsoPartyId, payload.round?.dso, payload.round?.provider];
	return candidates.some((value) => typeof value === "string" && value.trim() === target);
}

const utilityInstrumentConfigurationCache = new Map();
const utilityAllocationFactoryCache = new Map();
const utilityCredentialCache = new Map();
let cachedInstrumentConfigurationContracts = null;
let cachedAllocationFactoryContracts = null;
let featuredAppRightLedgerCache = null;
// Cache for consolidated Amulet context from transfer-factory endpoint
let amuletTransferFactoryCache = null;
let amuletTransferFactoryCacheExpiry = 0;
const AMULET_TRANSFER_FACTORY_CACHE_TTL_MS = 5 * 60 * 1000; // 5 minutes

function extractInstrumentIdentifier(payload) {
	return payload?.defaultIdentifier || payload?.instrument || payload?.instrumentId || null;
}

function matchesInstrumentConfiguration(payload, utilityAsset, registrar) {
	const identifier = extractInstrumentIdentifier(payload) || {};
	const assetId = identifier.id || payload?.instrumentId?.id || payload?.instrument?.id;
	const source = identifier.source || payload?.registrar || payload?.provider;
	const assetMatches = utilityAsset ? assetId === utilityAsset : true;
	const registrarMatches = registrar ? source === registrar || payload?.registrar === registrar || payload?.provider === registrar : true;
	return assetMatches && registrarMatches;
}

function matchesAllocationFactory(payload, registrar) {
	if (!registrar) {
		return true;
	}
	return payload?.registrar === registrar || payload?.operator === registrar || payload?.provider === registrar;
}

function extractCredentialHolder(payload) {
	return payload?.holder || payload?.credential?.holder || payload?.owner || payload?.party;
}

async function loadUtilityInstrumentConfigurationsCache(forceReload = false) {
	if (cachedInstrumentConfigurationContracts && !forceReload) {
		return cachedInstrumentConfigurationContracts;
	}
	const response = await getUtilityInstrumentConfigurations();
	const contracts = unwrapActiveContracts(response).map(mapActiveContractMetadata).filter(Boolean);
	cachedInstrumentConfigurationContracts = contracts;
	return contracts;
}

async function loadUtilityAllocationFactoriesCache(forceReload = false) {
	if (cachedAllocationFactoryContracts && !forceReload) {
		return cachedAllocationFactoryContracts;
	}
	const response = await getUtilityAllocationFactory();
	const contracts = unwrapActiveContracts(response).map(mapActiveContractMetadata).filter(Boolean);
	cachedAllocationFactoryContracts = contracts;
	return contracts;
}

export async function resolveUtilityInstrumentConfiguration(utilityAsset, registrar) {
	const cacheKey = `${registrar || "*"}::${utilityAsset || "*"}`;
	if (utilityInstrumentConfigurationCache.has(cacheKey)) {
		return utilityInstrumentConfigurationCache.get(cacheKey);
	}
	const contracts = await loadUtilityInstrumentConfigurationsCache();

	const match =
		contracts.find((entry) => matchesInstrumentConfiguration(entry.payload, utilityAsset, registrar)) ||
		contracts.find((entry) => matchesInstrumentConfiguration(entry.payload, utilityAsset, null)) ||
		null;
	const metadata = toContractMetadata(match);
	utilityInstrumentConfigurationCache.set(cacheKey, metadata);
	return metadata;
}

export async function resolveUtilityAllocationFactory(registrar, holder = null) {
	const cacheKey = `${registrar || "*"}::${holder || "*"}`;
	if (utilityAllocationFactoryCache.has(cacheKey)) {
		return utilityAllocationFactoryCache.get(cacheKey);
	}
	const contracts = await loadUtilityAllocationFactoriesCache();

	// If holder is specified, try to find an exact match first
	let match = null;
	if (holder) {
		match = contracts.find((entry) => matchesAllocationFactory(entry.payload, registrar) && entry.payload?.holder === holder);
	}

	// Fall back to registrar-only match if no holder-specific match found
	if (!match) {
		match = contracts.find((entry) => matchesAllocationFactory(entry.payload, registrar)) || null;
	}

	if (!match) {
		console.error(`No allocation factory found for registrar ${registrar}.`);
		return null;
	}
	const metadata = toContractMetadata(match);
	utilityAllocationFactoryCache.set(cacheKey, metadata);
	return metadata;
}

/**
 * Create a utility credential for a holder on a given registrar.
 * @param {string} holder - The party ID of the credential holder
 * @param {string} registrar - The registrar party ID
 * @param {string} [credentialId="tSBC"] - The credential identifier
 * @param {boolean} [returnCommand=false] - If true, returns the raw command and endpoint instead of executing
 * @returns {Promise<Object|null>} The created credential data, or null on failure
 */
export async function createUtilityCredential(holder, registrar, credentialId = "tSBC", returnCommand = false) {
	if (!holder || !registrar) {
		console.error("Missing holder or registrar for createUtilityCredential");
		return null;
	}

	const schema = getSchema("createUtilityCredential");
	const templateVars = {
		"utility-app-credential-packageId": shouldUseLedgerForMetadata() ? "#utility-credential-v0" : config.PACKAGE_UTILITY_APP_CREDENTIAL_ID,
		registrar: registrar,
		holder: holder,
		credential_id: credentialId,
		temple: config.TEMPLE_PARTY_ID,
	};
	const command = replaceTemplateVariables(schema, templateVars);
	const endpoint = `${config.VALIDATOR_API_URL}/v2/commands/submit-and-wait`;

	if (returnCommand) {
		return { command, endpoint };
	}

	const headers = await buildHeaders();

	try {
		const response = await axios.post(endpoint, command, { headers });
		return response.data;
	} catch (error) {
		const errorData = error?.response?.data;
		const errorDetail = errorData ? JSON.stringify(errorData, null, 2) : error.message;
		console.error(`Error creating utility credential: ${errorDetail}`);
		return null;
	}
}

async function resolveUtilityCredentialForParty(party, registrar = null, autoCreate = true) {
	if (!party) {
		return null;
	}
	if (utilityCredentialCache.has(party)) {
		return utilityCredentialCache.get(party);
	}
	const response = await getUtilitySenderCredentials(party);
	const contracts = unwrapActiveContracts(response).map(mapActiveContractMetadata).filter(Boolean);
	let match = contracts.find((entry) => extractCredentialHolder(entry.payload) === party) || null;

	if (!match && autoCreate && registrar) {
		// Auto-create credential for this party
		// Auto-create credential for this party
		const createResult = await createUtilityCredential(party, registrar);
		if (createResult) {
			// Clear cache and re-fetch to get the newly created credential
			utilityCredentialCache.delete(party);
			const refreshedResponse = await getUtilitySenderCredentials(party);
			const refreshedContracts = unwrapActiveContracts(refreshedResponse).map(mapActiveContractMetadata).filter(Boolean);
			match = refreshedContracts.find((entry) => extractCredentialHolder(entry.payload) === party) || null;
		}
	}

	if (!match) {
		console.error(`No credential found for party ${party}.`);
		return null;
	}
	const metadata = toContractMetadata(match);
	utilityCredentialCache.set(party, metadata);
	return metadata;
}

async function resolveUtilityCip56Context({ party, registrar, utilityAsset }) {
	// Get the expected registrar from instrument definition - NO FALLBACKS for utility tokens
	const instrumentDef = resolveInstrumentDefinition(utilityAsset);
	const network = config.NETWORK;
	const requiresCredentials = instrumentDef?.requireCredentials !== false;
	// For utility tokens, the registrar MUST come from the instrument catalog
	const catalogRegistrar = instrumentDef?.[network].registrar || null;
	const expectedRegistrar = registrar || catalogRegistrar;

	if (!expectedRegistrar) {
		console.error(`No registrar defined for utility asset ${utilityAsset} in instrument catalog`);
		return null;
	}

	// Use the registrar from: 1. Passed in value, or 2. Instrument catalog - NO OTHER FALLBACKS
	const resolvedRegistrar = registrar || catalogRegistrar;

	// For testnet/mainnet, use hardcoded contracts from instrument catalog
	if (network !== NETWORK_LOCALHOST && instrumentDef) {
		// Get the network-specific contracts from the instrument catalog
		const networkContracts = instrumentDef[network]; // testnet or mainnet

		if (!networkContracts) {
			console.error(`No ${network} contracts defined for utility asset ${utilityAsset} in instrument catalog`);
			return null;
		}

		const {
			instrumentConfiguration: instConfig,
			allocationFactory: allocFactory,
			appRewardConfiguration: appRewardConfig,
			featuredAppRight: featuredAppRightConfig,
			synchronizerId: catalogSynchronizerId,
		} = networkContracts;

		if (!instConfig || !allocFactory) {
			console.error(`Missing instrumentConfiguration or allocationFactory for ${utilityAsset} on ${network}`);
			return null;
		}

		if (!appRewardConfig) {
			console.warn(`Missing appRewardConfiguration for ${utilityAsset} on ${network} - validators may not receive rewards`);
		}

		if (!featuredAppRightConfig) {
			console.warn(`Missing featuredAppRight for ${utilityAsset} on ${network} - validators may not receive rewards`);
		}

		// Extract package IDs from the catalog's templateIds to ensure consistency with event blobs
		const extractPackageIdFromTemplateId = (templateId) => {
			if (!templateId) return null;
			const parts = templateId.split(":");
			return parts.length >= 1 ? parts[0] : null;
		};

		const registryPackageId = extractPackageIdFromTemplateId(instConfig.templateId) || config.PACKAGE_UTILITY_APP_REGISTRY_ID;
		const registryAppPackageId = extractPackageIdFromTemplateId(allocFactory.templateId) || config.PACKAGE_UTILITY_APP_REGISTRY_APP_ID;

		const instrumentConfiguration = {
			contractCid: normalizeContractId(instConfig.contractId),
			disclosureCid: instConfig.eventBlob,
			synchronizerId: catalogSynchronizerId,
			templateId: instConfig.templateId,
		};

		const allocationFactory = {
			contractCid: normalizeContractId(allocFactory.contractId),
			disclosureCid: allocFactory.eventBlob,
			synchronizerId: catalogSynchronizerId,
			templateId: allocFactory.templateId,
		};

		// Build appRewardConfiguration if available
		const appRewardConfiguration = appRewardConfig
			? {
					contractCid: normalizeContractId(appRewardConfig.contractId),
					disclosureCid: appRewardConfig.eventBlob,
					synchronizerId: catalogSynchronizerId,
					templateId: appRewardConfig.templateId,
				}
			: null;

		// Build featuredAppRight from catalog (for USDCx, this is their FeaturedAppRight, not ours)
		const featuredAppRight = featuredAppRightConfig
			? {
					contractCid: normalizeContractId(featuredAppRightConfig.contractId),
					disclosureCid: featuredAppRightConfig.eventBlob || null,
					synchronizerId: catalogSynchronizerId,
					templateId: featuredAppRightConfig.templateId || null,
				}
			: null;

		// For testnet/mainnet, sender credentials need to be queried from ledger
		// since they are user-specific (the user's Credential contract)
		const senderCredentials = requiresCredentials ? await resolveUtilityCredentialForParty(party, expectedRegistrar) : null;

		// Build disclosed contracts array for the command
		const disclosedContracts = [
			{
				templateId: instConfig.templateId,
				contractId: instConfig.contractId,
				createdEventBlob: instConfig.eventBlob,
				synchronizerId: catalogSynchronizerId,
			},
			{
				templateId: allocFactory.templateId,
				contractId: allocFactory.contractId,
				createdEventBlob: allocFactory.eventBlob,
				synchronizerId: catalogSynchronizerId,
			},
		];

		// Add appRewardConfiguration to disclosures if available (required for validator rewards)
		if (appRewardConfig) {
			disclosedContracts.push({
				templateId: appRewardConfig.templateId,
				contractId: appRewardConfig.contractId,
				createdEventBlob: appRewardConfig.eventBlob,
				synchronizerId: catalogSynchronizerId,
			});
		}

		// Add FeaturedAppRight to disclosures if available (required for validator rewards)
		if (featuredAppRight) {
			disclosedContracts.push({
				templateId: featuredAppRight.templateId,
				contractId: featuredAppRight.contractCid,
				createdEventBlob: featuredAppRight.disclosureCid,
				synchronizerId: featuredAppRight.synchronizerId,
			});
		}

		return {
			registrar: resolvedRegistrar,
			packages: {
				registryApp: registryAppPackageId,
				registry: registryPackageId,
				credential: config.PACKAGE_UTILITY_APP_CREDENTIAL_ID,
			},
			contextKeys: {
				instrumentConfiguration: DEFAULT_UTILITY_CONTEXT_KEYS.instrumentConfiguration,
				senderCredentials: DEFAULT_UTILITY_CONTEXT_KEYS.senderCredentials,
			},
			instrumentConfiguration,
			allocationFactory,
			appRewardConfiguration,
			featuredAppRight,
			senderCredentials,
			requiresCredentials,
			disclosedContracts,
		};
	}

	// For localhost, use direct ledger queries
	// For order creation, we only need sender credentials (the user's credentials)
	// Receiver credentials are resolved at settlement time, not order creation
	// Pass party as holder to find the correct AllocationFactory for this holder
	const instrumentConfigurationPromise = resolveUtilityInstrumentConfiguration(utilityAsset, expectedRegistrar);
	const allocationFactoryPromise = resolveUtilityAllocationFactory(expectedRegistrar, party);
	const senderCredentialsPromise = requiresCredentials ? resolveUtilityCredentialForParty(party, expectedRegistrar) : Promise.resolve(null);

	const [instrumentConfiguration, allocationFactory, senderCredentials] = await Promise.all([instrumentConfigurationPromise, allocationFactoryPromise, senderCredentialsPromise]);

	return {
		registrar: resolvedRegistrar,
		packages: {
			registryApp: config.PACKAGE_UTILITY_APP_REGISTRY_APP_ID,
			registry: config.PACKAGE_UTILITY_APP_REGISTRY_ID,
			credential: config.PACKAGE_UTILITY_APP_CREDENTIAL_ID,
		},
		contextKeys: {
			instrumentConfiguration: DEFAULT_UTILITY_CONTEXT_KEYS.instrumentConfiguration,
			senderCredentials: DEFAULT_UTILITY_CONTEXT_KEYS.senderCredentials,
		},
		instrumentConfiguration,
		allocationFactory,
		senderCredentials,
		requiresCredentials,
	};
}

/**
 * Precisely sum decimal string amounts using BigInt to avoid floating point errors.
 * Canton uses Numeric 10 (up to 10 decimal places).
 * @param {string[]} amounts - Array of decimal string amounts (e.g., ["1.0000000000", "6.8100000000"])
 * @returns {string} Precise sum as a decimal string
 */
function sumDecimalStrings(amounts) {
	const PRECISION = 10;
	const MULTIPLIER = BigInt(10 ** PRECISION);
	let total = BigInt(0);
	for (const amt of amounts) {
		const str = String(amt || "0");
		const parts = str.split(".");
		const intPart = BigInt(parts[0] || "0");
		const fracStr = (parts[1] || "").padEnd(PRECISION, "0").slice(0, PRECISION);
		total += intPart * MULTIPLIER + BigInt(fracStr);
	}
	const intResult = total / MULTIPLIER;
	const fracResult = total % MULTIPLIER;
	return `${intResult}.${fracResult.toString().padStart(PRECISION, "0")}`;
}

function toScanStyleContract(entry) {
	if (!entry) {
		return null;
	}
	const contractId = normalizeContractId(entry.contractId || entry.contractCid);
	if (!contractId) {
		return null;
	}
	return {
		contract_id: contractId,
		created_event_blob: entry.createdEventBlob,
		payload: entry.payload || {},
		template_id: entry.templateId || undefined,
	};
}

function buildContractMetadata(contract, domainId) {
	if (!contract) {
		return null;
	}
	const contractCid = normalizeContractId(contract.contract_id || contract.contractId);
	if (!contractCid || !contract.created_event_blob) {
		return null;
	}
	return {
		contractCid,
		disclosureCid: contract.created_event_blob,
		synchronizerId: domainId || null,
		templateId: contract.template_id || null,
	};
}

function getFeaturedAppRightEnvMetadata() {
	if (config.CONTRACT_FEATURED_APP_ID && config.DISCLOSURE_FEATURED_APP && config.SYNCHRONIZER_FEATURED_APP) {
		return {
			contractCid: normalizeContractId(config.CONTRACT_FEATURED_APP_ID),
			disclosureCid: config.DISCLOSURE_FEATURED_APP,
			synchronizerId: config.SYNCHRONIZER_FEATURED_APP || null,
			templateId: config.TEMPLATE_ID_FEATURED_APP,
		};
	}
	return null;
}

async function resolveFeaturedAppRightMetadata() {
	// Skip ledger call when using wallet adapter without validator URL
	const hasLedgerAccess = config.VALIDATOR_API_URL && !getWalletAdapter();
	if (hasLedgerAccess && shouldUseLedgerForMetadata()) {
		if (featuredAppRightLedgerCache) {
			return featuredAppRightLedgerCache;
		}
		const ledgerResponse = await getFeaturedAppRight();
		const ledgerContract = ledgerResponse?.featured_app_right?.contract;
		if (ledgerContract) {
			const metadata = buildContractMetadata(ledgerContract, ledgerResponse.featured_app_right.domain_id);
			if (metadata) {
				featuredAppRightLedgerCache = metadata;
				return metadata;
			}
		}
	}

	const envMetadata = getFeaturedAppRightEnvMetadata();
	if (envMetadata) {
		return envMetadata;
	}

	console.error("Missing Featured App Right metadata. Provide environment overrides or ensure ledger access.");
	return null;
}

/**
 * Resolves all Amulet context from a single transfer-factory call.
 * This consolidates what was previously 3 separate API calls:
 * - getAmuletRules() -> /amulet-rules
 * - getCandidateMiningRoundContract() -> /open-and-issuing-mining-rounds
 * - resolveExternalAmuletRulesMetadata() -> /registry/transfer-instruction/v1/transfer-factory
 *
 * The transfer-factory endpoint returns disclosedContracts containing:
 * - AmuletRules
 * - OpenMiningRound
 * - ExternalPartyAmuletRules
 * Plus choiceContextData with context keys (open-round, amulet-rules)
 *
 * @param {Object} params
 * @param {string} params.investor - The investor party ID
 * @param {string[]} params.holdingIds - Array of holding contract IDs
 * @param {number} params.transferAmount - The transfer amount
 * @returns {Object|null} Consolidated Amulet context or null on failure
 */
export async function resolveAmuletContext({ investor, holdingIds = [], transferAmount = 0 }) {
	const dso = config.VALIDATOR_DSO_PARTY_ID;

	// For localhost, fall back to individual queries
	if (shouldUseLedgerForMetadata()) {
		return resolveAmuletContextFromLedger({ investor, holdingIds, transferAmount });
	}

	// Check cache (with TTL)
	const now = Date.now();
	if (amuletTransferFactoryCache && now < amuletTransferFactoryCacheExpiry) {
		return amuletTransferFactoryCache;
	}

	// Fetch from transfer-factory endpoint
	const normalizedHoldingIds = (holdingIds || []).map(normalizeContractId).filter((id) => typeof id === "string" && id.length > 0);

	let transferFactoryData;
	try {
		transferFactoryData = await getTransferFactory(investor, dso, transferAmount, normalizedHoldingIds);
	} catch (error) {
		console.error(`Failed to fetch transfer factory: ${error}`);
		return null;
	}

	if (!transferFactoryData?.choiceContext) {
		console.error("Transfer factory response missing choiceContext");
		return null;
	}

	const disclosed = transferFactoryData.choiceContext.disclosedContracts;
	if (!Array.isArray(disclosed) || disclosed.length === 0) {
		console.error("Transfer factory response has no disclosed contracts");
		return null;
	}

	// Extract AmuletRules
	const amuletRulesDisclosure = findDisclosedContract(disclosed, "Splice.AmuletRules:AmuletRules");
	if (!amuletRulesDisclosure) {
		console.error("AmuletRules not found in transfer factory disclosed contracts");
		return null;
	}

	// Extract OpenMiningRound
	const openMiningRoundDisclosure = findDisclosedContract(disclosed, "Splice.Round:OpenMiningRound");
	if (!openMiningRoundDisclosure) {
		console.error("OpenMiningRound not found in transfer factory disclosed contracts");
		return null;
	}

	// Extract ExternalPartyAmuletRules
	const externalAmuletRulesDisclosure = findDisclosedContract(disclosed, "Splice.ExternalPartyAmuletRules:ExternalPartyAmuletRules");
	if (!externalAmuletRulesDisclosure) {
		console.error("ExternalPartyAmuletRules not found in transfer factory disclosed contracts");
		return null;
	}

	// Extract context keys from choiceContextData
	const contextData = transferFactoryData.choiceContext.choiceContextData?.values || {};
	const amuletRulesContextKey = Object.keys(contextData).find((k) => k.includes("amulet-rules") || k.includes("amuletRules")) || DEFAULT_AMULET_CONTEXT_KEYS.amuletRules;
	const openRoundContextKey = Object.keys(contextData).find((k) => k.includes("open-round") || k.includes("openRound")) || DEFAULT_AMULET_CONTEXT_KEYS.openRound;

	// Get Featured App Right from env (hardcoded)
	const featuredAppRightMetadata = await resolveFeaturedAppRightMetadata();
	if (!featuredAppRightMetadata) {
		console.error("Unable to resolve Featured App Right metadata.");
		return null;
	}

	const result = {
		amuletRules: {
			contractCid: normalizeContractId(amuletRulesDisclosure.contractId),
			disclosureCid: amuletRulesDisclosure.createdEventBlob,
			synchronizerId: amuletRulesDisclosure.synchronizerId,
			templateId: amuletRulesDisclosure.templateId,
		},
		openMiningRound: {
			contractCid: normalizeContractId(openMiningRoundDisclosure.contractId),
			disclosureCid: openMiningRoundDisclosure.createdEventBlob,
			synchronizerId: openMiningRoundDisclosure.synchronizerId,
			templateId: openMiningRoundDisclosure.templateId,
		},
		externalAmuletRules: {
			contractCid: normalizeContractId(externalAmuletRulesDisclosure.contractId),
			disclosureCid: externalAmuletRulesDisclosure.createdEventBlob,
			synchronizerId: externalAmuletRulesDisclosure.synchronizerId,
			templateId: externalAmuletRulesDisclosure.templateId,
		},
		featuredAppRight: featuredAppRightMetadata,
		contextKeys: {
			amuletRules: amuletRulesContextKey,
			openRound: openRoundContextKey,
			featuredAppRight: DEFAULT_AMULET_CONTEXT_KEYS.featuredAppRight,
			expireLock: DEFAULT_AMULET_CONTEXT_KEYS.expireLock,
		},
		dso,
	};

	// Cache the result
	amuletTransferFactoryCache = result;
	amuletTransferFactoryCacheExpiry = now + AMULET_TRANSFER_FACTORY_CACHE_TTL_MS;

	return result;
}

/**
 * Fallback for localhost: resolve Amulet context from individual ledger queries
 */
async function resolveAmuletContextFromLedger({ investor, holdingIds, transferAmount }) {
	const dso = config.VALIDATOR_DSO_PARTY_ID;

	// Parallel fetch all required contracts
	const [amuletRulesData, miningRoundsData, externalAmuletRulesData, featuredAppRightMetadata] = await Promise.all([
		getAmuletRules(dso),
		getOpenMiningRounds(dso),
		getExternalAmuletRules(dso),
		resolveFeaturedAppRightMetadata(),
	]);

	if (!amuletRulesData?.amulet_rules?.contract) {
		console.error("Failed to get AmuletRules from ledger");
		return null;
	}

	if (!miningRoundsData?.open_mining_rounds?.length) {
		console.error("Failed to get OpenMiningRounds from ledger");
		return null;
	}

	if (!externalAmuletRulesData?.external_amulet_rules?.contract) {
		console.error("Failed to get ExternalPartyAmuletRules from ledger");
		return null;
	}

	if (!featuredAppRightMetadata) {
		console.error("Failed to get FeaturedAppRight metadata");
		return null;
	}

	const amuletRulesContract = amuletRulesData.amulet_rules.contract;
	const amuletRulesDomainId = amuletRulesData.amulet_rules.domain_id;

	// Find a currently open mining round
	const now = new Date();
	let selectedRound = null;
	for (const roundEntry of miningRoundsData.open_mining_rounds) {
		const round = roundEntry.contract;
		const opensAt = new Date(round.payload?.opensAt);
		const targetClosesAt = new Date(round.payload?.targetClosesAt);
		if (now >= opensAt && now <= targetClosesAt) {
			selectedRound = {
				contract: round,
				domain_id: roundEntry.domain_id,
			};
			break;
		}
	}

	if (!selectedRound) {
		console.error("No currently open mining round found");
		return null;
	}

	const externalRulesContract = externalAmuletRulesData.external_amulet_rules.contract;
	const externalRulesDomainId = externalAmuletRulesData.external_amulet_rules.domain_id;

	const contextKeys = amuletRulesData.context_keys || {};

	return {
		amuletRules: {
			contractCid: normalizeContractId(amuletRulesContract.contract_id),
			disclosureCid: amuletRulesContract.created_event_blob,
			synchronizerId: amuletRulesDomainId,
			templateId: amuletRulesContract.template_id,
		},
		openMiningRound: {
			contractCid: normalizeContractId(selectedRound.contract.contract_id),
			disclosureCid: selectedRound.contract.created_event_blob,
			synchronizerId: selectedRound.domain_id,
			templateId: selectedRound.contract.template_id,
		},
		externalAmuletRules: {
			contractCid: normalizeContractId(externalRulesContract.contract_id),
			disclosureCid: externalRulesContract.created_event_blob,
			synchronizerId: externalRulesDomainId,
			templateId: externalRulesContract.template_id,
		},
		featuredAppRight: featuredAppRightMetadata,
		contextKeys: {
			amuletRules: contextKeys.amuletRules || contextKeys.amulet_rules || DEFAULT_AMULET_CONTEXT_KEYS.amuletRules,
			openRound: contextKeys.openRound || contextKeys.open_round || DEFAULT_AMULET_CONTEXT_KEYS.openRound,
			featuredAppRight: contextKeys.featuredAppRight || contextKeys.featured_app_right || DEFAULT_AMULET_CONTEXT_KEYS.featuredAppRight,
			expireLock: contextKeys.expireLock || contextKeys.expire_lock || DEFAULT_AMULET_CONTEXT_KEYS.expireLock,
		},
		dso,
	};
}

// Schemas object for easy lookup
const schemas = {
	createOrderProposalAmulet: createOrderProposalAmuletSchema,
	createOrderProposalUtility: createOrderProposalUtilitySchema,
	getAmuletHoldings: getAmuletHoldingsSchema,
	getLockedAmuletHoldings: getLockedAmuletHoldingsSchema,
	getUtilityHoldings: getUtilityHoldingsSchema,
	getUtilityInstrumentConfigurations: getUtilityInstrumentConfigurationsSchema,
	getUtilityAllocationFactory: getUtilityAllocationFactorySchema,
	getUtilitySenderCredentials: getUtilitySenderCredentialsSchema,
	getOrderProposals: getOrderProposalsSchema,
	getOrders: getOrdersSchema,
	getTransferFactory: getTransferFactorySchema,
	executeTransferFactory: executeTransferFactorySchema,
	unlockAmulet: unlockAmuletSchema,
	createUtilityCredential: createUtilityCredentialSchema,
};

// Helper function to get a schema by name
function getSchema(schemaName) {
	// Return a copy of the schema to avoid mutations
	return JSON.parse(JSON.stringify(schemas[schemaName]));
}

// Helper function to replace template variables in a schema
function replaceTemplateVariables(schema, variables) {
	let schemaStr = JSON.stringify(schema);

	// Always replace the {{randomUUID}} placeholder
	variables.randomUUID = randomUUID();

	// Replace each variable in the format {{variableName}}
	for (const [key, value] of Object.entries(variables)) {
		// Escape special regex characters in the key (including hyphens)
		const escapedKey = key.replace(/[.*+?^${}()|[\]\\-]/g, "\\$&");

		// First, try to replace as a full value: "{{key}}" -> proper JSON value
		const fullValueRegex = new RegExp(`"\\{\\{${escapedKey}\\}\\}"`, "g");
		let jsonValue = JSON.stringify(value);
		schemaStr = schemaStr.replace(fullValueRegex, jsonValue);

		// Then, replace as part of a string: {{key}} -> string value (without outer quotes)
		const partialValueRegex = new RegExp(`\\{\\{${escapedKey}\\}\\}`, "g");
		// For string values, remove the outer quotes from JSON.stringify result
		const stringValue = typeof value === "string" ? value : String(value);
		schemaStr = schemaStr.replace(partialValueRegex, stringValue);
	}

	return JSON.parse(schemaStr);
}

/**
 * Determine if an order requires Amulet context based on the trading pair and side.
 * Amulet context is needed when selling Amulet (base) or buying with Amulet (quote).
 * @param {string} baseAsset - The base asset (e.g., "Amulet")
 * @param {string} quoteAsset - The quote asset (e.g., "USDCx")
 * @param {string} side - "Buy" or "Sell"
 * @returns {boolean|null} True if Amulet context is needed, false if not, null if invalid symbol
 */
export function checkAmuletContext(baseAsset, quoteAsset, side) {
	const symbolPair = supportedSymbols[`${baseAsset}/${quoteAsset}`];
	if (!symbolPair) {
		return null; // Invalid symbol
	}

	return (side === "Sell" && baseAsset === "Amulet") || (side === "Buy" && quoteAsset === "Amulet");
}

async function queryLedgerContractsByTemplate(templateId, returnCommand = false) {
	if (!templateId) {
		console.error("Template ID is required to query ledger contracts.");
		return null;
	}

	const ledgerEnd = await getLedgerEnd();
	if (ledgerEnd === null) {
		console.error(`Cannot retrieve ledger end. Aborting query for template ${templateId}.`);
		return null;
	}

	const command = {
		verbose: true,
		activeAtOffset: ledgerEnd,
		filter: {
			filtersByParty: {},
			filtersForAnyParty: {
				cumulative: [
					{
						identifierFilter: {
							TemplateFilter: {
								value: {
									templateId,
									includeCreatedEventBlob: true,
								},
							},
						},
					},
				],
			},
		},
	};

	const endpoint = `${config.VALIDATOR_API_URL}/v2/state/active-contracts`;
	if (returnCommand) {
		return { command, endpoint };
	}

	const headers = await buildHeaders();

	try {
		const response = await axios.post(endpoint, command, { headers });
		return unwrapActiveContracts(response.data).map(mapActiveContractMetadata).filter(Boolean);
	} catch (error) {
		console.error(`Error fetching contracts for template ${templateId}: ${error}`);
		return null;
	}
}

/**
 * Get the AmuletRules contract from the Scan API or local ledger.
 * @param {string} dso - The DSO party ID
 * @param {boolean} [returnCommand=false] - If true, returns the raw command/endpoint instead of executing
 * @returns {Promise<Object|null>} The AmuletRules contract data, or null on failure
 */
export async function getAmuletRules(dso, returnCommand = false) {
	if (!shouldUseLedgerForMetadata()) {
		const headers = await buildHeaders();
		const url = `${config.VALIDATOR_SCAN_API_URL}/amulet-rules?provider=${dso}`;
		if (returnCommand) {
			return { command: null, endpoint: url, method: "GET" };
		}
		try {
			const response = await axios.get(url, { headers });
			return response.data;
		} catch (error) {
			console.error(`Error fetching Amulet Transfer Rules: ${error}`);
			return null;
		}
	}

	const templateId = "#splice-amulet:Splice.AmuletRules:AmuletRules";
	if (returnCommand) {
		return queryLedgerContractsByTemplate(templateId, true);
	}

	const contracts = await queryLedgerContractsByTemplate(templateId);
	if (!Array.isArray(contracts) || contracts.length === 0) {
		console.error("No Amulet Rules contracts found on the ledger. Ensure the local sandbox has been bootstrapped.");
		return null;
	}

	const selected = contracts.find((entry) => matchesDso(entry.payload, dso)) || contracts[0];
	const contract = toScanStyleContract(selected);
	if (!contract) {
		console.error("Unable to format Amulet Rules contract from ledger response.");
		return null;
	}

	const contextKeys = buildAmuletContextKeys(selected.payload?.contextKeys || selected.payload?.context_keys || {});

	return {
		amulet_rules: {
			contract,
			domain_id: selected.synchronizerId,
		},
		context_keys: contextKeys,
	};
}

/**
 * Get the ExternalPartyAmuletRules contract from the local ledger.
 * @param {string} dso - The DSO party ID
 * @param {boolean} [returnCommand=false] - If true, returns the raw command/endpoint instead of executing
 * @returns {Promise<Object|null>} The ExternalPartyAmuletRules contract data, or null on failure
 */
export async function getExternalAmuletRules(dso, returnCommand = false) {
	const templateId = "#splice-amulet:Splice.ExternalPartyAmuletRules:ExternalPartyAmuletRules";
	if (returnCommand) {
		return queryLedgerContractsByTemplate(templateId, true);
	}

	const contracts = await queryLedgerContractsByTemplate(templateId);
	if (!Array.isArray(contracts) || contracts.length === 0) {
		console.error("No External Party Amulet Rules contracts found on the ledger. Ensure sandbox bootstrap completed.");
		return null;
	}

	const filtered = dso ? contracts.filter((entry) => matchesDso(entry.payload, dso)) : contracts;
	const selected = filtered[0] || contracts[0];
	const contract = toScanStyleContract(selected);
	if (!contract) {
		console.error("Unable to format External Party Amulet Rules contract from ledger response.");
		return null;
	}

	return {
		external_amulet_rules: {
			contract,
			domain_id: selected.synchronizerId,
		},
	};
}

/**
 * Get the FeaturedAppRight contract from the local ledger.
 * @param {boolean} [returnCommand=false] - If true, returns the raw command/endpoint instead of executing
 * @returns {Promise<Object|null>} The FeaturedAppRight contract data, or null on failure
 */
export async function getFeaturedAppRight(returnCommand = false) {
	const templateId = "#splice-amulet:Splice.Amulet:FeaturedAppRight";
	if (returnCommand) {
		return queryLedgerContractsByTemplate(templateId, true);
	}

	const contracts = await queryLedgerContractsByTemplate(templateId);
	if (!Array.isArray(contracts) || contracts.length === 0) {
		console.error("No Featured App Right contracts found on the ledger. Ensure the local environment has been initialized.");
		return null;
	}

	const selected = contracts[0];
	const contract = toScanStyleContract(selected);
	if (!contract) {
		console.error("Unable to format Featured App Right contract from ledger response.");
		return null;
	}

	return {
		featured_app_right: {
			contract,
			domain_id: selected.synchronizerId,
		},
	};
}

/**
 * Get open and issuing mining rounds from the Scan API or local ledger.
 * @param {string} dso - The DSO party ID
 * @param {boolean} [returnCommand=false] - If true, returns the raw command/endpoint instead of executing
 * @returns {Promise<Object|null>} Mining rounds data with open_mining_rounds array, or null on failure
 */
export async function getOpenMiningRounds(dso, returnCommand = false) {
	if (!shouldUseLedgerForMetadata()) {
		const headers = await buildHeaders();
		const url = `${config.VALIDATOR_SCAN_API_URL}/open-and-issuing-mining-rounds?provider=${dso}`;
		if (returnCommand) {
			return { command: null, endpoint: url, method: "GET" };
		}
		try {
			const response = await axios.get(url, { headers });
			return response.data;
		} catch (error) {
			console.error(`Error fetching open mining rounds: ${error}`);
			return null;
		}
	}

	const templateId = "#splice-amulet:Splice.Round:OpenMiningRound";
	if (returnCommand) {
		return queryLedgerContractsByTemplate(templateId, true);
	}
	const contracts = await queryLedgerContractsByTemplate(templateId);
	if (!Array.isArray(contracts) || contracts.length === 0) {
		console.error("No Open Mining Round contracts found on the ledger. Ensure the local sandbox has been bootstrapped.");
		return null;
	}

	const filtered = dso ? contracts.filter((entry) => matchesDso(entry.payload, dso)) : contracts;
	const source = filtered.length > 0 ? filtered : contracts;
	const sorted = source.slice().sort((a, b) => {
		const aTime = new Date(a.payload?.opensAt || a.payload?.openAt || 0).getTime();
		const bTime = new Date(b.payload?.opensAt || b.payload?.openAt || 0).getTime();
		return bTime - aTime;
	});
	const openRounds = sorted
		.map((entry) => ({
			contract: toScanStyleContract(entry),
			domain_id: entry.synchronizerId,
		}))
		.filter((entry) => entry.contract !== null);

	return {
		open_mining_rounds: openRounds,
		issuing_mining_rounds: [],
	};
}

// Build transfer factory data for AMULET from local ledger (for localhost when Scan API is unavailable)
async function buildLocalAmuletTransferFactory(investor, dso, amount, holdingIds) {
	// Get ExternalPartyAmuletRules - this is the factory contract for Amulet
	const externalRulesData = await getExternalAmuletRules(dso);
	if (!externalRulesData || !externalRulesData.external_amulet_rules) {
		console.error("Failed to get ExternalPartyAmuletRules from local ledger");
		return null;
	}
	const externalRules = externalRulesData.external_amulet_rules;
	const factoryContractId = normalizeContractId(externalRules.contract.contract_id);

	// Get AmuletRules
	const amuletRulesData = await getAmuletRules(dso);
	if (!amuletRulesData || !amuletRulesData.amulet_rules) {
		console.error("Failed to get AmuletRules from local ledger");
		return null;
	}
	const amuletRules = amuletRulesData.amulet_rules;
	const amuletRulesContractId = normalizeContractId(amuletRules.contract.contract_id);

	// Get OpenMiningRound
	const miningRound = await getCandidateMiningRoundContract(dso);
	if (!miningRound) {
		console.error("Failed to get OpenMiningRound from local ledger");
		return null;
	}
	const openRoundContractId = normalizeContractId(miningRound.contractId);

	// Build the choiceContextData structure
	const choiceContextData = {
		values: {
			"open-round": {
				tag: "AV_ContractId",
				value: openRoundContractId,
			},
			"amulet-rules": {
				tag: "AV_ContractId",
				value: amuletRulesContractId,
			},
		},
	};

	// Build disclosed contracts array - use template_id from the actual contracts when available
	const disclosedContracts = [
		{
			templateId: externalRules.contract.template_id || "#splice-amulet:Splice.ExternalPartyAmuletRules:ExternalPartyAmuletRules",
			contractId: factoryContractId,
			createdEventBlob: externalRules.contract.created_event_blob,
			synchronizerId: externalRules.domain_id,
		},
		{
			templateId: amuletRules.contract.template_id || "#splice-amulet:Splice.AmuletRules:AmuletRules",
			contractId: amuletRulesContractId,
			createdEventBlob: amuletRules.contract.created_event_blob,
			synchronizerId: amuletRules.domain_id,
		},
		{
			templateId: miningRound.templateId || "#splice-amulet:Splice.Round:OpenMiningRound",
			contractId: openRoundContractId,
			createdEventBlob: miningRound.createdEventBlob,
			synchronizerId: miningRound.synchronizerId,
		},
	];

	return {
		factoryId: factoryContractId,
		choiceContext: {
			choiceContextData: choiceContextData,
			disclosedContracts: disclosedContracts,
		},
		transfer: {
			amount: amount,
			instrumentId: {
				admin: dso,
				id: "Amulet",
			},
		},
		expectedAdmin: dso,
	};
}

// Build transfer factory data for UTILITY TOKENS from local ledger
async function buildLocalUtilityTransferFactory(investor, registrar, amount, holdingIds, utilityAsset) {
	// Get the full utility context - includes AllocationFactory, InstrumentConfiguration, and Credentials
	const utilityContext = await resolveUtilityCip56Context({
		party: investor,
		registrar: registrar,
		utilityAsset: utilityAsset,
	});

	if (!utilityContext) {
		console.error(`Failed to get utility context for registrar ${registrar} from local ledger`);
		return null;
	}

	const { allocationFactory, instrumentConfiguration, senderCredentials, packages, contextKeys, requiresCredentials } = utilityContext;

	if (!allocationFactory?.contractCid) {
		console.error(`Failed to get AllocationFactory for registrar ${registrar} from local ledger`);
		return null;
	}

	const factoryContractId = normalizeContractId(allocationFactory.contractCid);

	// Build the choiceContextData structure for utility tokens
	// Utility tokens always need instrument-configuration in the context and
	// only include sender-credentials when the registrar requires them
	const choiceContextData = {
		values: {},
	};

	// Add instrument configuration to context if available
	if (instrumentConfiguration?.contractCid && contextKeys?.instrumentConfiguration) {
		choiceContextData.values[contextKeys.instrumentConfiguration] = {
			tag: "AV_ContractId",
			value: normalizeContractId(instrumentConfiguration.contractCid),
		};
	}

	// Add sender credentials to context when required and available
	// Note: sender-credentials must be a LIST of contract IDs
	if (requiresCredentials && senderCredentials?.contractCid && contextKeys?.senderCredentials) {
		choiceContextData.values[contextKeys.senderCredentials] = {
			tag: "AV_List",
			value: [
				{
					tag: "AV_ContractId",
					value: normalizeContractId(senderCredentials.contractCid),
				},
			],
		};
	}

	// Build disclosed contracts array - include all required contracts
	const disclosedContracts = [];

	// AllocationFactory (required)
	if (allocationFactory?.contractCid && allocationFactory?.disclosureCid) {
		disclosedContracts.push({
			templateId: `${packages?.registryApp || config.PACKAGE_UTILITY_APP_REGISTRY_APP_ID}:Utility.Registry.App.V0.Service.AllocationFactory:AllocationFactory`,
			contractId: factoryContractId,
			createdEventBlob: allocationFactory.disclosureCid,
			synchronizerId: allocationFactory.synchronizerId,
		});
	}

	// InstrumentConfiguration (required for transfers)
	if (instrumentConfiguration?.contractCid && instrumentConfiguration?.disclosureCid) {
		disclosedContracts.push({
			templateId: `${packages?.registry || config.PACKAGE_UTILITY_APP_REGISTRY_ID}:Utility.Registry.V0.Configuration.Instrument:InstrumentConfiguration`,
			contractId: normalizeContractId(instrumentConfiguration.contractCid),
			createdEventBlob: instrumentConfiguration.disclosureCid,
			synchronizerId: instrumentConfiguration.synchronizerId,
		});
	}

	// SenderCredentials (only when required)
	if (requiresCredentials && senderCredentials?.contractCid && senderCredentials?.disclosureCid) {
		disclosedContracts.push({
			templateId: `${packages?.credential || config.PACKAGE_UTILITY_APP_CREDENTIAL_ID}:Utility.Credential.V0.Credential:Credential`,
			contractId: normalizeContractId(senderCredentials.contractCid),
			createdEventBlob: senderCredentials.disclosureCid,
			synchronizerId: senderCredentials.synchronizerId,
		});
	}

	return {
		factoryId: factoryContractId,
		choiceContext: {
			choiceContextData: choiceContextData,
			disclosedContracts: disclosedContracts,
		},
		transfer: {
			amount: amount,
			instrumentId: {
				admin: registrar,
				id: utilityAsset || "tSBC",
			},
		},
		expectedAdmin: registrar,
	};
}

// Build transfer factory data from local ledger (for localhost when Scan API is unavailable)
// Determines whether to use Amulet or Utility token path based on the admin party
async function buildLocalTransferFactory(investor, admin, amount, holdingIds, utilityAsset = null) {
	// Determine if this is an Amulet (DSO) or utility token (Registrar) operation
	const isAmulet = admin && admin.includes("DSO");

	if (isAmulet) {
		return buildLocalAmuletTransferFactory(investor, admin, amount, holdingIds);
	} else {
		return buildLocalUtilityTransferFactory(investor, admin, amount, holdingIds, utilityAsset);
	}
}

// Get transfer factory from the public registry API (no auth required).
// Used by wallet provider flows where Scan API is not available.
async function getTransferFactoryFromRegistry(investor, registrar, amount, holdingIds, utilityAsset) {
	const instrumentDef = resolveInstrumentDefinition(utilityAsset);
	const networkContracts = instrumentDef?.[config.NETWORK];
	const registryAPI = networkContracts?.registryAPI;
	if (!registryAPI) {
		console.error(`No registryAPI defined for ${utilityAsset} on ${config.NETWORK}`);
		return null;
	}
	const url = `${registryAPI}/transfer-instruction/v1/transfer-factory`;
	const schema = getSchema("getTransferFactory");
	const templateVars = {
		investor: investor,
		dso: registrar,
		amount: amount,
		holding_ids: holdingIds,
		instrument_id: utilityAsset,
		time: new Date().toISOString(),
		timeEnd: new Date(Date.now() + 3600 * 1000).toISOString(),
	};
	const body = replaceTemplateVariables(schema, templateVars);
	try {
		const response = await axios.post(url, body);
		return response.data;
	} catch (error) {
		console.error(`Error fetching transfer factory from registry API: ${error}`);
		if (error.response) {
			console.error(`Response data: ${JSON.stringify(error.response.data)}`);
		}
		return null;
	}
}

/**
 * Get a transfer factory from the Scan API, or build one locally for localhost environments.
 * @param {string} investor - The investor party ID
 * @param {string} admin - The admin party ID (DSO for Amulet, registrar for utility tokens)
 * @param {number|string} amount - The transfer amount
 * @param {string[]} holdingIds - Array of holding contract IDs
 * @param {boolean} [returnCommand=false] - If true, returns the raw command/endpoint instead of executing
 * @param {string|null} [utilityAsset=null] - The utility asset ID (e.g., "USDCx") for localhost utility transfers
 * @returns {Promise<Object|null>} Transfer factory data, or null on failure
 */
export async function getTransferFactory(investor, admin, amount, holdingIds, returnCommand = false, utilityAsset = null) {
	// For localhost, build the transfer factory data from local ledger contracts
	if (shouldUseLedgerForMetadata()) {
		if (returnCommand) {
			// Return a placeholder indicating this will be built at runtime
			return { command: null, endpoint: "localhost-ledger", method: "LOCAL" };
		}
		return buildLocalTransferFactory(investor, admin, amount, holdingIds, utilityAsset);
	}

	const headers = await buildHeaders();
	const url = `${config.VALIDATOR_SCAN_API_URL}/registry/transfer-instruction/v1/transfer-factory`;
	const schema = getSchema("getTransferFactory");
	const templateVars = {
		investor: investor,
		dso: admin,
		amount: amount,
		holding_ids: holdingIds,
		instrument_id: utilityAsset || "Amulet",
		time: new Date().toISOString(),
		timeEnd: new Date(Date.now() + 3600 * 1000).toISOString(),
	};
	const command = replaceTemplateVariables(schema, templateVars);

	if (returnCommand) {
		return { command, endpoint: url };
	}

	try {
		const response = await axios.post(url, command, { headers });
		return response.data;
	} catch (error) {
		console.error(`Error fetching transfer factory: ${error}`);
		return null;
	}
}

/**
 * Get Amulet disclosure contracts (AmuletRules, ExternalPartyAmuletRules, OpenMiningRound) from the Scan API.
 * Used by external wallet integrations that cannot access the Scan API directly.
 * Not available for localhost environments.
 * @param {string} party - The party ID to fetch disclosures for
 * @returns {Promise<Array|null>} Array of disclosed contract objects, or null on failure
 */
export async function getAmuletDisclosures(party) {
	// For localhost, this endpoint is not available
	if (shouldUseLedgerForMetadata()) {
		console.error("getAmuletDisclosures is not available for localhost - use ledger-based resolution instead");
		return null;
	}

	const headers = await buildHeaders();
	const url = `${config.VALIDATOR_SCAN_API_URL}/registry/transfer-instruction/v1/transfer-factory`;
	const dso = config.VALIDATOR_DSO_PARTY_ID;

	// Use amount=0 and empty holding IDs to just get the disclosures without creating a real transfer
	const now = new Date();
	const tomorrow = new Date(now.getTime() + 24 * 60 * 60 * 1000);

	const requestBody = {
		choiceArguments: {
			expectedAdmin: dso,
			transfer: {
				sender: party,
				receiver: party,
				amount: "0",
				instrumentId: {
					admin: dso,
					id: "Amulet",
				},
				requestedAt: now.toISOString(),
				executeBefore: tomorrow.toISOString(),
				inputHoldingCids: [],
				meta: { values: {} },
			},
			extraArgs: {
				context: { values: {} },
				meta: { values: {} },
			},
		},
		excludeDebugFields: true,
	};

	try {
		const response = await axios.post(url, requestBody, { headers });

		// Return the disclosed contracts array
		if (response.data && response.data?.choiceContext?.disclosedContracts) {
			return response.data.choiceContext.disclosedContracts;
		}

		console.error("No disclosedContracts found in transfer-factory response");
		return null;
	} catch (error) {
		console.error(`Error fetching amulet disclosures: ${error}`);
		if (error.response) {
			console.error(`Response data: ${JSON.stringify(error.response.data)}`);
		}
		return null;
	}
}

/**
 * Get a currently-open mining round contract suitable for use in transactions.
 * @param {string} dso - The DSO party ID
 * @param {boolean} [returnCommand=false] - If true, returns the raw command/endpoint instead of executing
 * @returns {Promise<Object|null>} Mining round metadata (contractId, createdEventBlob, synchronizerId), or null if none found
 */
export async function getCandidateMiningRoundContract(dso, returnCommand = false) {
	const miningRoundsData = await getOpenMiningRounds(dso);
	if (!miningRoundsData || !miningRoundsData.open_mining_rounds || miningRoundsData.open_mining_rounds.length === 0) {
		console.error(`No open mining rounds found for DSO ${dso}`);
		return null;
	}

	const now = new Date();

	// Find a mining round that is currently open
	for (const roundEntry of miningRoundsData.open_mining_rounds) {
		const round = roundEntry.contract;
		const opensAt = new Date(round.payload.opensAt);
		const targetClosesAt = new Date(round.payload.targetClosesAt);

		if (now >= opensAt && now <= targetClosesAt) {
			return {
				contractId: round.contract_id,
				createdEventBlob: round.created_event_blob,
				synchronizerId: roundEntry.domain_id,
				templateId: round.template_id || null,
			};
		}
	}

	console.error(`No currently open mining round found for DSO ${dso}`);
	return null;
}

/**
 * Get Amulet holdings for a party from the ledger or a provider.
 * @param {string} party - The party ID
 * @param {boolean} [returnCommand=false] - If true, returns the raw command/endpoint instead of executing
 * @param {Object|null} [provider=null] - Optional external provider (e.g., Loop Wallet) to use instead of ledger API
 * @returns {Promise<Object|null>} Holdings data, or null on failure
 */
export async function getAmuletHoldingsForParty(party, returnCommand = false, provider = null) {
	provider = resolveProvider(provider);
	// When using provider, skip ledger API calls entirely
	if (provider) {
		const args = {
			templateId: "#splice-amulet:Splice.Amulet:Amulet",
		};
		try {
			return await provider.getActiveContracts(args);
		} catch (error) {
			console.error(`Error fetching Amulet holdings via provider: ${error}`);
			return null;
		}
	}

	// Only call ledger API when NOT using provider
	const ledgerEnd = await getLedgerEnd();
	if (ledgerEnd === null) {
		console.error("Cannot retrieve ledger end. Aborting getAmuletHoldingsForParty.");
		return null;
	}

	// Get schema and replace template variables
	const schema = getSchema("getAmuletHoldings");
	const templateVars = {
		ledger_end: ledgerEnd,
		investor: party,
	};
	const command = replaceTemplateVariables(schema, templateVars);

	const endpoint = `${config.VALIDATOR_API_URL}/v2/state/active-contracts`;

	if (returnCommand) {
		return { command, endpoint };
	}

	const headers = await buildHeaders();
	try {
		const response = await axios.post(endpoint, command, { headers });
		return response.data;
	} catch (error) {
		console.error(`Error fetching Amulet holdings: ${error}`);
		return null;
	}
}

/**
 * Get locked Amulet holdings for a party from the ledger or a provider.
 * @param {string} party - The party ID
 * @param {boolean} [returnCommand=false] - If true, returns the raw command/endpoint instead of executing
 * @param {Object|null} [provider=null] - Optional external provider to use instead of ledger API
 * @returns {Promise<Object|null>} Locked holdings data, or null on failure
 */
export async function getLockedAmuletHoldingsForParty(party, returnCommand = false, provider = null) {
	provider = resolveProvider(provider);
	// When using provider, skip ledger API calls entirely
	if (provider) {
		const args = {
			templateId: "#splice-amulet:Splice.Amulet:LockedAmulet",
		};
		try {
			return await provider.getActiveContracts(args);
		} catch (error) {
			console.error(`Error fetching locked Amulet holdings via provider: ${error}`);
			return null;
		}
	}

	const ledgerEnd = await getLedgerEnd();
	if (ledgerEnd === null) {
		console.error("Cannot retrieve ledger end. Aborting getLockedAmuletHoldingsForParty.");
		return null;
	}

	// Get schema and replace template variables
	const schema = getSchema("getLockedAmuletHoldings");
	const templateVars = {
		ledger_end: ledgerEnd,
		investor: party,
	};
	const command = replaceTemplateVariables(schema, templateVars);

	if (returnCommand) {
		return { command, endpoint: `${config.VALIDATOR_API_URL}/v2/state/active-contracts` };
	}

	const headers = await buildHeaders();
	const endpoint = `${config.VALIDATOR_API_URL}/v2/state/active-contracts`;
	try {
		const response = await axios.post(endpoint, command, { headers });
		return response.data;
	} catch (error) {
		console.error(`Error fetching Amulet holdings: ${error}`);
		return null;
	}
}

/**
 * Get utility token holdings for a party from the ledger or a provider.
 * @param {string} party - The party ID
 * @param {boolean} [returnCommand=false] - If true, returns the raw command/endpoint instead of executing
 * @param {Object|null} [provider=null] - Optional external provider to use instead of ledger API
 * @returns {Promise<Object|null>} Holdings data, or null on failure
 */
export async function getUtilityHoldingsForParty(party, returnCommand = false, provider = null) {
	provider = resolveProvider(provider);
	// When using provider, skip ledger API calls entirely
	if (provider) {
		const args = {
			templateId: "#utility-registry-holding-v0:Utility.Registry.Holding.V0.Holding:Holding",
		};
		try {
			return await provider.getActiveContracts(args);
		} catch (error) {
			console.error(`Error fetching Utility holdings via provider: ${error}`);
			return null;
		}
	}

	// Only call ledger API when NOT using provider
	const ledgerEnd = await getLedgerEnd();
	if (ledgerEnd === null) {
		console.error("Cannot retrieve ledger end. Aborting getUtilityHoldingsForParty.");
		return null;
	}

	// Get schema and replace template variables
	const schema = getSchema("getUtilityHoldings");
	const templateVars = {
		ledger_end: ledgerEnd,
		investor: party,
	};
	const command = replaceTemplateVariables(schema, templateVars);
	const endpoint = `${config.VALIDATOR_API_URL}/v2/state/active-contracts`;

	if (returnCommand) {
		return { command, endpoint };
	}

	const headers = await buildHeaders();
	try {
		const response = await axios.post(endpoint, command, { headers });
		return response.data;
	} catch (error) {
		console.error(`Error fetching Utility holdings: ${error}`);
		return null;
	}
}

/**
 * Get UTXO summary for a given asset and party.
 * Useful for confirming a merge was successful and checking if a holding is large enough for an order.
 * @param {string} party - The party ID
 * @param {string} assetId - The asset ID (e.g., "USDCx", "CBTC", "Amulet")
 * @param {Object|null} [provider=null] - Optional wallet provider
 * @returns {Promise<{total: number, unlocked: number, locked: number, largestUnlocked: number, unlockedBalance: number}>} UTXO summary
 */
export async function getUtxoCount(party, assetId, provider = null) {
	provider = resolveProvider(provider);
	const isAmulet = assetId === "Amulet";
	const holdings = isAmulet ? await getAmuletHoldingsForParty(party, false, provider) : await getUtilityHoldingsForParty(party, false, provider);

	if (!holdings || holdings.length === 0) {
		return { total: 0, unlocked: 0, locked: 0, largestUnlocked: 0, unlockedBalance: 0 };
	}

	let unlocked = 0;
	let locked = 0;
	let largestUnlocked = 0;
	let unlockedBalance = 0;

	for (const h of holdings) {
		const createdEvent = h.contractEntry?.JsActiveContract?.createdEvent;
		if (!createdEvent) continue;
		const createArg = createdEvent.createArgument;

		let isLocked = false;
		let quantity = 0;
		let matchesAsset = false;

		if (isAmulet) {
			isLocked = !!(createArg?.lock || createArg?.amulet?.lock);
			quantity = parseFloat(createArg?.amount?.initialAmount || 0);
			matchesAsset = true;
		} else {
			const asset = createArg?.instrument?.id || null;
			if (asset !== assetId) continue;
			isLocked = !!createArg?.lock;
			quantity = parseFloat(createArg?.amount || 0);
			matchesAsset = true;
		}

		if (!matchesAsset) continue;

		if (isLocked) {
			locked++;
		} else {
			unlocked++;
			unlockedBalance += quantity;
			if (quantity > largestUnlocked) {
				largestUnlocked = quantity;
			}
		}
	}

	return { total: unlocked + locked, unlocked, locked, largestUnlocked, unlockedBalance };
}

/**
 * Retrieve utility instrument configuration contracts from the ledger.
 * @param {boolean} [returnCommand=false] - If true, returns the raw command/endpoint instead of executing
 * @returns {Promise<Object|null>} Instrument configuration contracts, or null on failure
 */
export async function getUtilityInstrumentConfigurations(returnCommand = false) {
	const ledgerEnd = await getLedgerEnd();
	if (ledgerEnd === null) {
		console.error("Cannot retrieve ledger end. Aborting getUtilityInstrumentConfigurations.");
		return null;
	}

	const schema = getSchema("getUtilityInstrumentConfigurations");
	const templateVars = {
		ledger_end: ledgerEnd,
		utility_app_registry_package_id: shouldUseLedgerForMetadata() ? "#utility-registry-v0" : config.PACKAGE_UTILITY_APP_REGISTRY_ID,
	};
	const command = replaceTemplateVariables(schema, templateVars);
	const endpoint = `${config.VALIDATOR_API_URL}/v2/state/active-contracts`;

	if (returnCommand) {
		return { command, endpoint };
	}

	const headers = await buildHeaders();

	try {
		const response = await axios.post(endpoint, command, { headers });
		return response.data;
	} catch (error) {
		console.error(`Error fetching utility instrument configurations: ${error}`);
		return null;
	}
}

/**
 * Retrieve the utility allocation factory contract from the ledger.
 * @param {boolean} [returnCommand=false] - If true, returns the raw command/endpoint instead of executing
 * @returns {Promise<Object|null>} Allocation factory contract data, or null on failure
 */
export async function getUtilityAllocationFactory(returnCommand = false) {
	const ledgerEnd = await getLedgerEnd();
	if (ledgerEnd === null) {
		console.error("Cannot retrieve ledger end. Aborting getUtilityAllocationFactory.");
		return null;
	}

	const schema = getSchema("getUtilityAllocationFactory");
	const templateVars = {
		ledger_end: ledgerEnd,
		utility_app_registry_app_package_id: shouldUseLedgerForMetadata() ? "#utility-registry-app-v0" : config.PACKAGE_UTILITY_APP_REGISTRY_APP_ID,
	};
	const command = replaceTemplateVariables(schema, templateVars);
	const endpoint = `${config.VALIDATOR_API_URL}/v2/state/active-contracts`;

	if (returnCommand) {
		return { command, endpoint };
	}

	const headers = await buildHeaders();

	try {
		const response = await axios.post(endpoint, command, { headers });
		return response.data;
	} catch (error) {
		console.error(`Error fetching utility allocation factory: ${error}`);
		return null;
	}
}

/**
 * Retrieve utility sender credentials for a party from the ledger.
 * @param {string} sender - The sender party ID
 * @param {boolean} [returnCommand=false] - If true, returns the raw command/endpoint instead of executing
 * @returns {Promise<Object|null>} Sender credentials data, or null on failure
 */
export async function getUtilitySenderCredentials(sender, returnCommand = false) {
	const ledgerEnd = await getLedgerEnd();
	if (ledgerEnd === null) {
		console.error("Cannot retrieve ledger end. Aborting getUtilitySenderCredentials.");
		return null;
	}

	if (!sender) {
		console.error("Missing sender party identifier for getUtilitySenderCredentials.");
		return null;
	}

	const schema = getSchema("getUtilitySenderCredentials");
	const templateVars = {
		ledger_end: ledgerEnd,
		sender: sender,
		"utility-app-credential-packageId": shouldUseLedgerForMetadata() ? "#utility-credential-v0" : config.PACKAGE_UTILITY_APP_CREDENTIAL_ID,
	};
	const command = replaceTemplateVariables(schema, templateVars);
	const endpoint = `${config.VALIDATOR_API_URL}/v2/state/active-contracts`;

	if (returnCommand) {
		return { command, endpoint };
	}

	const headers = await buildHeaders();

	try {
		const response = await axios.post(endpoint, command, { headers });
		return response.data;
	} catch (error) {
		console.error(`Error fetching utility sender credentials: ${error}`);
		return null;
	}
}

/**
 * Find a suitable holding contract for order creation that meets the required quantity.
 * Searches Amulet or utility holdings and validates operator/provider/registrar compatibility.
 * @param {string} party - The party ID
 * @param {boolean} isAmulet - True to search Amulet holdings, false for utility holdings
 * @param {number|string} requiredQuantity - The minimum quantity required
 * @param {string|null} [expectedAssetId=null] - Expected asset ID for utility holdings (e.g., "USDCx")
 * @param {boolean} [returnCommand=false] - If true, returns the raw command/endpoint instead of executing
 * @param {Object|null} [provider=null] - Optional external provider to use instead of ledger API
 * @returns {Promise<Object>} Holding data with contractId, quantity, dso/registrar, or an error object
 */
export async function getCandidateHoldingForOrderCreation(party, isAmulet, requiredQuantity, expectedAssetId = null, returnCommand = false, provider = null) {
	provider = resolveProvider(provider);
	const holdings = isAmulet ? await getAmuletHoldingsForParty(party, false, provider) : await getUtilityHoldingsForParty(party, false, provider);

	if (!holdings || holdings.length === 0) {
		console.error(`No holdings found for party ${party}`);
		return {
			error: `No holdings found for party ${party}`,
		};
	}

	const network = config.NETWORK;

	// For utility holdings, resolve the expected values from the instrument definition
	let expectedRegistrar = null;
	let expectedInstrumentScheme = null;
	let expectedOperator = null;
	let expectedProvider = null;
	if (!isAmulet && expectedAssetId) {
		const instrumentDef = resolveInstrumentDefinition(expectedAssetId);
		if (instrumentDef) {
			expectedRegistrar = instrumentDef[network].registrar || null;
			expectedInstrumentScheme = instrumentDef[network].scheme || RegistrarInternalScheme;
			expectedOperator = instrumentDef[network].operator || null;
			expectedProvider = instrumentDef[network].provider || null;
		}
	}

	// Extract the actual contract data and map to a simpler structure
	const contracts = holdings
		.map((h) => {
			const createdEvent = h.contractEntry?.JsActiveContract?.createdEvent;
			if (!createdEvent) return null;
			const synchronizerId = h.contractEntry?.JsActiveContract?.synchronizerId;

			// For Amulet, DSO is in createArgument.dso
			// For Utility, registrar is in createArgument.instrument.source (which is the registrar)
			const holdingRegistrar = isAmulet ? null : createdEvent.createArgument?.registrar || createdEvent.createArgument?.instrument?.source || null;
			const dso = isAmulet ? createdEvent.createArgument?.dso || null : holdingRegistrar;

			// For Amulet, quantity is in amount.initialAmount
			// For Utility, quantity is in createArgument.amount
			const quantity = isAmulet ? parseFloat(createdEvent.createArgument?.amount?.initialAmount || 0) : parseFloat(createdEvent.createArgument?.amount || 0);

			// For Utility, also get the asset ID and scheme
			const asset = isAmulet ? "Amulet" : createdEvent.createArgument?.instrument?.id || null;
			const instrumentScheme = isAmulet ? null : createdEvent.createArgument?.instrument?.scheme || null;

			// Extract owner/holder/operator/provider for debugging
			const owner = createdEvent.createArgument?.owner || null;
			const holder = createdEvent.createArgument?.holder || null;
			const operator = createdEvent.createArgument?.operator || null;
			const provider = createdEvent.createArgument?.provider || null;

			// For Utility check the lock status and skip locked holdings. Amulets are never locked here.
			if (!isAmulet && createdEvent.createArgument?.lock) {
				return null;
			}

			return {
				contractId: normalizeContractId(createdEvent.contractId),
				quantity: quantity,
				createdEventBlob: createdEvent.createdEventBlob,
				synchronizerId: synchronizerId,
				dso: dso,
				registrar: holdingRegistrar,
				asset: asset,
				instrumentScheme: instrumentScheme,
				owner: owner,
				holder: holder,
				operator: operator,
				provider: provider,
			};
		})
		.filter((c) => c !== null);

	if (contracts.length === 0) {
		console.error(`No valid holdings found for party ${party}`);
		return {
			error: `No valid holdings found for party ${party}`,
		};
	}

	// For utility tokens, filter by expected registrar and asset if provided
	let eligibleContracts = contracts;
	if (!isAmulet) {
		eligibleContracts = contracts.filter((holding) => {
			// Filter by owner - the holding MUST belong to the party creating the order
			if (holding.owner && holding.owner !== party) {
				return false;
			}
			// Filter by expected asset if provided
			if (expectedAssetId && holding.asset) {
				if (holding.asset.toLowerCase() !== expectedAssetId.toLowerCase()) {
					return false;
				}
			}
			// Filter by expected registrar - MUST match for utility tokens
			if (expectedRegistrar && holding.registrar) {
				if (holding.registrar.toLowerCase() !== expectedRegistrar.toLowerCase()) {
					return false;
				}
			}
			// Filter by operator/provider - holdings must match the AllocationFactory's operator/provider
			// These values come from the instrument catalog to ensure compatibility
			if (expectedOperator && holding.operator && holding.operator !== expectedOperator) {
				return false;
			}
			if (expectedProvider && holding.provider && holding.provider !== expectedProvider) {
				return false;
			}
			return true;
		});

		if (eligibleContracts.length === 0) {
			console.error(
				`No holdings found for party ${party} matching owner=${party}, asset=${expectedAssetId}, registrar=${expectedRegistrar}, operator=${expectedOperator}, provider=${expectedProvider}`,
			);
			console.error(`Holdings available but filtered out may have incompatible operator/provider`);
			// Do NOT fall back - mismatch will cause transaction failure
			return {
				error: `No holdings found for party ${party} matching owner=${party}, asset=${expectedAssetId}, registrar=${expectedRegistrar}, operator=${expectedOperator}, provider=${expectedProvider}`,
			};
		}
	}

	// Sort by quantity ascending
	const sortedHoldings = eligibleContracts.sort((a, b) => a.quantity - b.quantity);

	// Check if any holding has sufficient quantity
	let totalBalance = 0;

	for (const holding of sortedHoldings) {
		totalBalance += holding.quantity;

		if (holding.quantity >= requiredQuantity) {
			// For Amulets, check DSO matches
			// For Utility tokens, DSO check is not applicable (source is different)
			if (isAmulet) {
				if (holding.dso === config.VALIDATOR_DSO_PARTY_ID) {
					return holding;
				} else {
					console.warn(`Holding ${holding.contractId} has sufficient quantity but DSO ${holding.dso} does not match required DSO ${config.VALIDATOR_DSO_PARTY_ID}`);
				}
			} else {
				return holding;
			}
		}
	}

	// No single holding has sufficient quantity
	console.error(`No single holding found for party ${party} with required quantity ${requiredQuantity}. Total balance available: ${totalBalance}`);
	console.error(`User has enough total balance but needs to merge his holdings`);

	return {
		error: `User has enough total balance but needs to merge his holdings`,
		data: {
			totalBalance: totalBalance,
			requiredQuantity: requiredQuantity,
		},
	};
}

/**
 * Merge Amulet holdings for a party into a single holding via a transfer factory.
 * Requires at least 2 holdings to merge. When using a provider (wallet integration),
 * returns the command for the wallet to sign instead of submitting directly.
 * @param {string} party - The party ID
 * @param {boolean} [returnCommand=false] - If true, returns the raw command/endpoint instead of executing
 * @param {Object|null} [provider=null] - Optional external provider (e.g., Loop Wallet) to use instead of ledger API
 * @param {number|null} [maxUtxos=null] - Maximum number of UTXOs to merge (picks smallest first). Null = merge all.
 * @param {Array} [amuletDisclosures=[]] - Pre-fetched Amulet disclosures (required when using a provider). Must contain AmuletRules, OpenMiningRound, and ExternalPartyAmuletRules.
 * @returns {Promise<Object|null>} Merge result data, or null on failure
 */
export async function mergeAmuletHoldingsForParty(party, returnCommand = false, provider = null, maxUtxos = null, amuletDisclosures = []) {
	provider = resolveProvider(provider);
	if (provider) {
		if (!amuletDisclosures || amuletDisclosures.length === 0) {
			console.error("amuletDisclosures are required when using a provider for Amulet merge");
			return { error: "amuletDisclosures are required when using a provider for Amulet merge" };
		}
		returnCommand = true;
		// Wallet providers (e.g., Loop) have a ~100 contract limit per transaction.
		if (!maxUtxos) {
			maxUtxos = 100;
		}
	}

	if (!provider) {
		const ledgerEnd = await getLedgerEnd();
		if (ledgerEnd === null) {
			console.error("Cannot retrieve ledger end. Aborting mergeAmuletHoldingsForParty.");
			return null;
		}
	}

	// To merge amulets we need to transfer them to the same party with a combined output of the total amount

	const holdings = await getAmuletHoldingsForParty(party, false, provider);
	if (!holdings || holdings.length < 2) {
		console.error(`Not enough holdings to merge for party ${party}`);
		return null;
	}

	// Extract the actual contract data and map to a simpler structure
	const contracts = holdings
		.map((h) => {
			const createdEvent = h.contractEntry?.JsActiveContract?.createdEvent;
			if (!createdEvent) return null;
			const synchronizerId = h.contractEntry?.JsActiveContract?.synchronizerId;
			const dso = createdEvent.createArgument?.dso || createdEvent.createArgument?.amulet?.dso || null;
			const rawAmount = createdEvent.createArgument?.amount?.initialAmount || createdEvent.createArgument?.amulet?.amount?.initialAmount || "0";
			const quantity = parseFloat(rawAmount);

			const contractId = normalizeContractId(createdEvent.contractId);

			return {
				contractId,
				quantity: quantity,
				rawAmount: rawAmount,
				createdEventBlob: createdEvent.createdEventBlob,
				synchronizerId: synchronizerId,
				dso: dso,
			};
		})
		.filter((c) => c !== null);

	if (contracts.length < 2) {
		console.error(`Not enough valid holdings to merge for party ${party}`);
		return null;
	}

	// Sort by quantity ascending (smallest first) and apply maxUtxos limit
	contracts.sort((a, b) => a.quantity - b.quantity);
	const selectedContracts = maxUtxos && maxUtxos < contracts.length ? contracts.slice(0, maxUtxos) : contracts;

	if (selectedContracts.length < 2) {
		console.error(`Not enough holdings after applying maxUtxos limit for party ${party}`);
		return null;
	}

	// Calculate total quantity using precise decimal math (BigInt) to avoid floating point drift
	const totalQuantity = sumDecimalStrings(selectedContracts.map((c) => c.rawAmount));

	// For simplicity, use the first holding's DSO and synchronizer
	const primaryHolding = selectedContracts[0];
	if (!primaryHolding.dso) {
		console.error(`Primary holding ${primaryHolding.contractId} does not have a DSO associated.`);
		return null;
	}

	// Get the transfer factory for the merge
	const holdingIds = selectedContracts.map((c) => c.contractId);
	const normalizedHoldingIds = normalizeContractReferences(holdingIds);

	let transferFactoryData;
	if (provider) {
		// Build transfer factory from pre-fetched amuletDisclosures
		const findDisclosure = (suffix) => amuletDisclosures.find((d) => d.templateId?.includes(suffix) || d.template_id?.includes(suffix));
		const amuletRulesDisclosure = findDisclosure("Splice.AmuletRules:AmuletRules");
		const openMiningRoundDisclosure = findDisclosure("Splice.Round:OpenMiningRound");
		const externalAmuletRulesDisclosure = findDisclosure("Splice.ExternalPartyAmuletRules:ExternalPartyAmuletRules");

		if (!amuletRulesDisclosure || !openMiningRoundDisclosure || !externalAmuletRulesDisclosure) {
			console.error("Missing required disclosures in amuletDisclosures array. Need AmuletRules, OpenMiningRound, and ExternalPartyAmuletRules.");
			return { error: "Missing required disclosures in amuletDisclosures array. Need AmuletRules, OpenMiningRound, and ExternalPartyAmuletRules." };
		}

		const dso = primaryHolding.dso;
		const factoryCid = normalizeContractId(externalAmuletRulesDisclosure.contractId || externalAmuletRulesDisclosure.contract_id);
		const amuletRulesCid = normalizeContractId(amuletRulesDisclosure.contractId || amuletRulesDisclosure.contract_id);
		const openRoundCid = normalizeContractId(openMiningRoundDisclosure.contractId || openMiningRoundDisclosure.contract_id);

		transferFactoryData = {
			factoryId: factoryCid,
			choiceContext: {
				choiceContextData: {
					values: {
						"open-round": { tag: "AV_ContractId", value: openRoundCid },
						"amulet-rules": { tag: "AV_ContractId", value: amuletRulesCid },
					},
				},
				disclosedContracts: [
					{
						templateId: externalAmuletRulesDisclosure.templateId || externalAmuletRulesDisclosure.template_id,
						contractId: factoryCid,
						createdEventBlob: externalAmuletRulesDisclosure.createdEventBlob || externalAmuletRulesDisclosure.created_event_blob,
						synchronizerId:
							externalAmuletRulesDisclosure.synchronizerId ||
							externalAmuletRulesDisclosure.synchronizer_id ||
							externalAmuletRulesDisclosure.domainId ||
							externalAmuletRulesDisclosure.domain_id,
					},
					{
						templateId: amuletRulesDisclosure.templateId || amuletRulesDisclosure.template_id,
						contractId: amuletRulesCid,
						createdEventBlob: amuletRulesDisclosure.createdEventBlob || amuletRulesDisclosure.created_event_blob,
						synchronizerId: amuletRulesDisclosure.synchronizerId || amuletRulesDisclosure.synchronizer_id || amuletRulesDisclosure.domainId || amuletRulesDisclosure.domain_id,
					},
					{
						templateId: openMiningRoundDisclosure.templateId || openMiningRoundDisclosure.template_id,
						contractId: openRoundCid,
						createdEventBlob: openMiningRoundDisclosure.createdEventBlob || openMiningRoundDisclosure.created_event_blob,
						synchronizerId:
							openMiningRoundDisclosure.synchronizerId || openMiningRoundDisclosure.synchronizer_id || openMiningRoundDisclosure.domainId || openMiningRoundDisclosure.domain_id,
					},
				],
			},
			transfer: {
				amount: totalQuantity,
				instrumentId: { admin: dso, id: "Amulet" },
			},
			expectedAdmin: dso,
		};
	} else {
		// Pass 0 as amount to let the factory calculate the total effective amount
		transferFactoryData = await getTransferFactory(party, primaryHolding.dso, 0, normalizedHoldingIds);
	}

	if (!transferFactoryData || !transferFactoryData.choiceContext) {
		console.error(`Failed to get transfer factory for merging holdings of party ${party}`);
		return null;
	}

	// Extract the factory contract ID and choice context
	const factoryContractId = normalizeContractId(transferFactoryData.factoryId);
	const choiceContextRaw = transferFactoryData.choiceContext.choiceContextData;
	const disclosedContractsRaw = transferFactoryData.choiceContext.disclosedContracts || [];
	const choiceContext = normalizeContractReferences(choiceContextRaw);
	const disclosedContracts = normalizeContractReferences(disclosedContractsRaw);

	// Get the actual transfer metadata from the factory response (accounts for holding fees)
	const transferInstruction = transferFactoryData.transfer || {};
	// Format amount to avoid floating point precision issues (Numeric 10 max precision)
	const rawAmount = transferInstruction.amount || totalQuantity;
	const actualAmount = parseFloat(parseFloat(rawAmount).toFixed(10));
	const instrumentAdmin = transferInstruction.instrumentId?.admin || primaryHolding.dso;
	const instrumentIdValue = transferInstruction.instrumentId?.id || "Amulet";
	const expectedAdmin = transferFactoryData.expectedAdmin || instrumentAdmin;

	// Build the transfer command using the transfer factory
	const schema = getSchema("executeTransferFactory");
	const now = new Date();
	// requestedAt MUST be in the past for Canton transfers
	const requestedAt = new Date(now.getTime() - 60 * 1000); // 1 minute in the past
	// executeBefore should be well in the future to allow processing time
	const executeBefore = new Date(now.getTime() + 3600 * 1000); // 1 hour in the future

	const templateVars = {
		factory_contract_id: factoryContractId,
		expected_admin: expectedAdmin,
		instrument_admin: instrumentAdmin,
		instrument_id: instrumentIdValue,
		investor: party,
		amount: actualAmount.toString(),
		holding_ids: normalizedHoldingIds,
		time: requestedAt.toISOString(),
		timeEnd: executeBefore.toISOString(),
		choice_context: choiceContext,
		disclosed_contracts: disclosedContracts,
		user_id: config.AUTH0_USER_ID || party,
		randomUUID: randomUUID(),
	};

	const command = replaceTemplateVariables(schema, templateVars);

	if (returnCommand) {
		return { command, endpoint: `${config.VALIDATOR_API_URL}/v2/commands/submit-and-wait` };
	}

	const headers = await buildHeaders();
	const endpoint = `${config.VALIDATOR_API_URL}/v2/commands/submit-and-wait`;

	try {
		const response = await axios.post(endpoint, command, { headers });
		// Merge successful
		return response.data;
	} catch (error) {
		console.error(`Error executing transfer factory for merge: ${error}`);
		if (error.response) {
			console.error(`Response data: ${JSON.stringify(error.response.data)}`);
		}
		return null;
	}
}

/**
 * Merge all utility token holdings of a specific asset for a party into a single holding.
 * Requires at least 2 holdings to merge.
 * @param {string} party - The party ID
 * @param {string} utilityAsset - The utility asset ID (e.g., "USDCx", "CBTC")
 * @param {boolean} [returnCommand=false] - If true, returns the raw command/endpoint instead of executing. Forced to true when provider is set.
 * @param {Object} [provider=null] - Wallet provider (e.g., Loop Wallet) with getActiveContracts(). When set, skips ledger API calls and returns the command for the wallet to sign.
 * @param {number} [maxUtxos=null] - Maximum number of UTXOs to merge. When set, picks the smallest holdings first.
 * @returns {Promise<Object|null>} Merge result data, or {command, endpoint} when returnCommand is true, or null on failure
 */
export async function mergeUtilityHoldingsForParty(party, utilityAsset, returnCommand = false, provider = null, maxUtxos = null) {
	provider = resolveProvider(provider);
	if (provider) {
		returnCommand = true;
		// Wallet providers (e.g., Loop) have a ~100 contract limit per transaction.
		// Default to 100 when no explicit maxUtxos is set.
		if (!maxUtxos) {
			maxUtxos = 100;
		}
	}

	if (!provider) {
		const ledgerEnd = await getLedgerEnd();
		if (ledgerEnd === null) {
			console.error("Cannot retrieve ledger end. Aborting mergeUtilityHoldingsForParty.");
			return null;
		}
	}

	// To merge utility tokens we need to transfer them to the same party with a combined output of the total amount

	const holdings = await getUtilityHoldingsForParty(party, false, provider);
	if (!holdings || holdings.length < 2) {
		console.error(`Not enough holdings to merge for party ${party}`);
		return null;
	}

	// Extract the actual contract data and map to a simpler structure, filtering by utility asset
	const contracts = holdings
		.map((h) => {
			const createdEvent = h.contractEntry?.JsActiveContract?.createdEvent;
			if (!createdEvent) return null;
			const synchronizerId = h.contractEntry?.JsActiveContract?.synchronizerId;
			const createArg = createdEvent.createArgument;

			const source = createArg?.instrument?.source || null;
			const rawAmount = createArg?.amount || "0";
			const quantity = parseFloat(rawAmount);
			const asset = createArg?.instrument?.id || null;
			const lock = createArg?.lock || false;
			// Extract the holder from the holding contract (for utility tokens)
			// Try multiple possible field names
			const holder = createArg?.holder || createArg?.owner || createArg?.account?.owner || null;
			// Extract operator, provider, registrar from holding for group identifier
			const operator = createArg?.operator || null;
			const provider = createArg?.provider || null;
			const registrar = createArg?.registrar || null;

			if (lock || asset !== utilityAsset) {
				return null; // Skip locked holdings or different asset types
			}
			const contractId = normalizeContractId(createdEvent.contractId);
			return {
				contractId,
				templateId: createdEvent.templateId,
				quantity: quantity,
				rawAmount: rawAmount,
				createdEventBlob: createdEvent.createdEventBlob,
				synchronizerId: synchronizerId,
				source: source,
				asset: asset,
				holder: holder,
				operator: operator,
				provider: provider,
				registrar: registrar,
			};
		})
		.filter((c) => c !== null && c.asset === utilityAsset);

	if (contracts.length < 2) {
		console.error(`Not enough valid ${utilityAsset} holdings to merge for party ${party}. Found ${contracts.length} holdings.`);
		return null;
	}

	// Filter holdings to ensure they all have the same operator/provider group
	// The AllocationFactory expects all inputs to belong to the same group identifier
	const uniqueOperators = [...new Set(contracts.map((c) => c.operator).filter(Boolean))];
	const uniqueProviders = [...new Set(contracts.map((c) => c.provider).filter(Boolean))];

	// If we have holdings with different operators/providers, filter to the most common group
	if (uniqueOperators.length > 1 || uniqueProviders.length > 1) {
		console.warn(`Found holdings with mixed operator/provider groups. Filtering to match Temple operator...`);
		// Count holdings per operator to find the dominant group
		const operatorCounts = {};
		contracts.forEach((c) => {
			const op = c.operator || "unknown";
			operatorCounts[op] = (operatorCounts[op] || 0) + 1;
		});
		// Prefer Temple operator, or fall back to most common
		const templeOperator = Object.keys(operatorCounts).find((op) => op.startsWith("Temple::"));
		const targetOperator = templeOperator || Object.keys(operatorCounts).sort((a, b) => operatorCounts[b] - operatorCounts[a])[0];

		// Filter to only holdings with the target operator
		const filteredContracts = contracts.filter((c) => c.operator === targetOperator);
		// Filtered holdings to match target operator

		if (filteredContracts.length < 2) {
			console.error(`Not enough holdings with matching operator after filtering. Found ${filteredContracts.length} holdings.`);
			return null;
		}

		// Replace contracts with filtered list
		contracts.length = 0;
		contracts.push(...filteredContracts);
	}

	// Filter by holder/owner - only merge holdings belonging to the same party
	// The requesting party parameter should match the holder field in holdings
	const uniqueHolders = [...new Set(contracts.map((c) => c.holder).filter(Boolean))];
	if (uniqueHolders.length > 1) {
		console.warn(`Found holdings for multiple holders (${uniqueHolders.join(", ")}). Filtering to match requesting party...`);

		// Try to find holdings that match the requesting party
		let targetHolder = uniqueHolders.find((h) => h === party || h.startsWith(party.split("::")[0] + "::"));

		if (!targetHolder) {
			// If no exact match, prefer the first holder (most holdings)
			const holderCounts = {};
			contracts.forEach((c) => {
				const h = c.holder || "unknown";
				holderCounts[h] = (holderCounts[h] || 0) + 1;
			});
			targetHolder = Object.keys(holderCounts).sort((a, b) => holderCounts[b] - holderCounts[a])[0];
		}

		const filteredByHolder = contracts.filter((c) => c.holder === targetHolder);

		if (filteredByHolder.length < 2) {
			console.error(`Not enough holdings with matching holder after filtering. Found ${filteredByHolder.length} holdings.`);
			return null;
		}

		// Replace contracts with filtered list
		contracts.length = 0;
		contracts.push(...filteredByHolder);
	}

	// Recalculate unique holders after filtering
	const finalHolders = [...new Set(contracts.map((c) => c.holder).filter(Boolean))];
	const holder = finalHolders[0] || party; // Use the holder from filtered holdings

	// Sort ascending by quantity so smallest UTXOs are merged first
	if (maxUtxos) {
		contracts.sort((a, b) => a.quantity - b.quantity);
		if (contracts.length > maxUtxos) {
			contracts.length = maxUtxos;
		}
		if (contracts.length < 2) {
			console.error(`Not enough holdings after applying maxUtxos limit (${maxUtxos}). Found ${contracts.length}.`);
			return null;
		}
	}

	// Calculate total quantity using precise decimal math (BigInt) to avoid floating point drift
	const totalQuantity = sumDecimalStrings(contracts.map((c) => c.rawAmount));

	// Gather holding contract IDs to merge
	const holdingIds = contracts.map((c) => c.contractId);
	const normalizedHoldingIds = normalizeContractReferences(holdingIds);

	const uniqueSources = [...new Set(contracts.map((c) => c.source).filter(Boolean))];
	if (uniqueSources.length === 0) {
		console.error(`Unable to resolve registrar for ${utilityAsset} holdings; merger aborted.`);
		return null;
	}
	if (uniqueSources.length > 1) {
		console.error(`Found holdings for multiple registrars (${uniqueSources.join(", ")}); merge requires a single registrar context.`);
		return null;
	}

	const registrar = uniqueSources[0];

	// Resolve transfer factory:
	// - Localhost: build from local ledger contracts (handles both Amulet and utility)
	// - Remote: use the instrument's Registry API (Scan API only handles Amulet)
	let transferFactoryData;
	if (shouldUseLedgerForMetadata()) {
		transferFactoryData = await getTransferFactory(holder, registrar, totalQuantity, normalizedHoldingIds, false, utilityAsset);
	} else {
		transferFactoryData = await getTransferFactoryFromRegistry(holder, registrar, totalQuantity, normalizedHoldingIds, utilityAsset);
	}
	if (!transferFactoryData || !transferFactoryData.choiceContext) {
		console.error(`Failed to retrieve transfer factory context for ${utilityAsset} holdings of party ${party}`);
		return null;
	}

	const factoryContractId = normalizeContractId(transferFactoryData.factoryId);
	if (!factoryContractId) {
		console.error(`Transfer factory response missing factoryId for ${utilityAsset} holdings.`);
		return null;
	}

	const transferInstruction = transferFactoryData.transfer || {};
	// Use factory amount if available, fall back to our precise sum
	const rawAmount = transferInstruction.amount || totalQuantity;
	const actualAmount = parseFloat(parseFloat(rawAmount).toFixed(10));
	const instrumentAdmin = transferInstruction.instrumentId?.admin || registrar;
	const instrumentIdValue = transferInstruction.instrumentId?.id || utilityAsset;
	// For utility tokens the TransferFactory is registrar-administered, so expectedAdmin = registrar.
	// Prefer the value returned by the Registry API, then fall back to the registrar from holdings.
	const expectedAdmin = transferFactoryData.expectedAdmin || registrar;

	const choiceContextRaw = transferFactoryData.choiceContext.choiceContextData || {};
	const disclosedContractsRaw = transferFactoryData.choiceContext.disclosedContracts || [];
	const choiceContext = normalizeContractReferences(choiceContextRaw);
	let disclosedContracts = normalizeContractReferences(disclosedContractsRaw);

	// Add input holdings as disclosed contracts - required for AllocationFactory to verify group identifiers
	// The holdings need to be disclosed for the factory to see their content
	const holdingDisclosures = contracts
		.filter((c) => c.createdEventBlob && c.contractId)
		.map((c) => ({
			templateId: c.templateId || `${config.PACKAGE_UTILITY_APP_ID}:Utility.Registry.Holding.V0.Holding:Holding`,
			contractId: normalizeContractId(c.contractId),
			createdEventBlob: c.createdEventBlob,
			synchronizerId: c.synchronizerId || config.GLOBAL_SYNCHRONIZER_ID,
		}));

	if (holdingDisclosures.length > 0) {
		// Add holding disclosures to transfer command
		disclosedContracts = [...disclosedContracts, ...holdingDisclosures];
	}

	const now = new Date();
	// requestedAt MUST be in the past for Canton transfers
	const requestedAt = new Date(now.getTime() - 60 * 1000); // 1 minute in the past
	// executeBefore should be well in the future to allow processing time
	const executeBefore = new Date(now.getTime() + 3600 * 1000); // 1 hour in the future

	const schema = getSchema("executeTransferFactory");
	const templateVars = {
		factory_contract_id: factoryContractId,
		expected_admin: expectedAdmin,
		instrument_admin: instrumentAdmin,
		instrument_id: instrumentIdValue,
		investor: holder, // Use holder from holdings, not the query party
		amount: actualAmount.toString(),
		holding_ids: normalizedHoldingIds,
		time: requestedAt.toISOString(),
		timeEnd: executeBefore.toISOString(),
		choice_context: choiceContext,
		disclosed_contracts: disclosedContracts,
		user_id: config.AUTH0_USER_ID || party,
		randomUUID: randomUUID(),
	};

	const command = replaceTemplateVariables(schema, templateVars);

	if (returnCommand) {
		return { command, endpoint: `${config.VALIDATOR_API_URL}/v2/commands/submit-and-wait` };
	}

	const headers = await buildHeaders();
	const endpoint = `${config.VALIDATOR_API_URL}/v2/commands/submit-and-wait`;

	try {
		const response = await axios.post(endpoint, command, { headers });
		// Utility merge successful
		return response.data;
	} catch (error) {
		console.error(`Error executing transfer factory for ${utilityAsset} merge: ${error}`);
		if (error.response) {
			console.error(`Response data: ${JSON.stringify(error.response.data)}`);
		}
		return null;
	}
}

/**
 * Split an Amulet holding into multiple holdings with specified output quantities.
 * @param {string} party - The party ID
 * @param {number[]} outputQuantity - Array of desired output quantities
 * @param {boolean} [returnCommand=false] - If true, returns the raw command/endpoint instead of executing
 * @returns {Promise<Object|null>} Split result data, or null on failure
 */
export async function splitAmuletHoldingForParty(party, outputQuantity, returnCommand = false) {
	// Splitting amulets is similar to merging but in reverse
	// We transfer from one holding into multiple holdings with specified output quantities

	const holding = await getCandidateHoldingForOrderCreation(party, true, outputQuantity);
	if (!holding || holding.error) {
		console.error(`No suitable holding found for party ${party} to split.`);
		return null;
	}

	// Get the transfer factory for the split
	const holdingIds = [holding.contractId];
	const normalizedHoldingIds = normalizeContractReferences(holdingIds);
	// Pass 0 as amount to let the factory calculate the total effective amount
	const transferFactoryData = await getTransferFactory(party, holding.dso, 0, normalizedHoldingIds);
	if (!transferFactoryData || !transferFactoryData.choiceContext) {
		console.error(`Failed to get transfer factory for splitting holding of party ${party}`);
		return null;
	}

	// Extract the factory contract ID and choice context
	const factoryContractId = normalizeContractId(transferFactoryData.factoryId);
	const choiceContextRaw = transferFactoryData.choiceContext.choiceContextData;
	const disclosedContractsRaw = transferFactoryData.choiceContext.disclosedContracts || [];
	const choiceContext = normalizeContractReferences(choiceContextRaw);
	const disclosedContracts = normalizeContractReferences(disclosedContractsRaw);

	// Get the actual transfer amount from the factory response (accounts for holding fees)
	const transferInstruction = transferFactoryData.transfer || {};
	const actualAmount = transferInstruction.amount || outputQuantity;
	const instrumentAdmin = transferInstruction.instrumentId?.admin || holding.dso;
	const instrumentIdValue = transferInstruction.instrumentId?.id || "Amulet";
	const expectedAdmin = transferInstruction.instrumentId?.admin || transferFactoryData.expectedAdmin || holding.dso;

	// Build the transfer command using the transfer factory
	const schema = getSchema("executeTransferFactory");
	const now = new Date();
	// Setting executeBefore to 30 seconds from now to account for processing time
	const executeBefore = new Date(now.getTime() + 30 * 1000);

	const templateVars = {
		factory_contract_id: factoryContractId,
		expected_admin: expectedAdmin,
		instrument_admin: instrumentAdmin,
		instrument_id: instrumentIdValue,
		investor: party,
		amount: actualAmount.toString(),
		holding_ids: normalizedHoldingIds,
		time: now.toISOString(),
		timeEnd: executeBefore.toISOString(),
		choice_context: choiceContext,
		disclosed_contracts: disclosedContracts,
		user_id: config.AUTH0_USER_ID,
		randomUUID: randomUUID(),
	};

	const command = replaceTemplateVariables(schema, templateVars);

	if (returnCommand) {
		return { command, endpoint: `${config.VALIDATOR_API_URL}/v2/commands/submit-and-wait` };
	}

	const headers = await buildHeaders();
	const endpoint = `${config.VALIDATOR_API_URL}/v2/commands/submit-and-wait`;

	try {
		const response = await axios.post(endpoint, command, { headers });
		// Split successful
		return response.data;
	} catch (error) {
		console.error(`Error executing transfer factory for split: ${error}`);
		if (error.response) {
			console.error(`Response data: ${JSON.stringify(error.response.data)}`);
		}
		return null;
	}
}

/**
 * Unlock all locked Amulet holdings for a party.
 * @param {string} party - The party ID
 * @param {boolean} [returnCommand=false] - If true, returns the raw commands instead of executing
 * @returns {Promise<Array|null>} Array of unlock results, or null if no locked amulets found
 */
export async function unlockLockedAmulets(party, returnCommand = false) {
	const lockedHoldings = await getLockedAmuletHoldingsForParty(party);
	if (!lockedHoldings || lockedHoldings.length === 0) {
		// No locked amulets found
		return null;
	}

	// Process locked amulets

	// Get the amulet rules first
	const amuletRulesData = await getAmuletRules(config.VALIDATOR_DSO_PARTY_ID);
	if (!amuletRulesData || !amuletRulesData.amulet_rules || Object.keys(amuletRulesData.amulet_rules).length === 0) {
		console.error(`No Amulet Transfer Rules found for DSO ${config.VALIDATOR_DSO_PARTY_ID}`);
		return null;
	}

	// Get a candidate mining round
	const miningRound = await getCandidateMiningRoundContract(config.VALIDATOR_DSO_PARTY_ID);
	if (!miningRound) {
		console.error(`No open Mining Round found for DSO ${config.VALIDATOR_DSO_PARTY_ID}`);
		return null;
	}

	const featuredAppRightMetadata = await resolveFeaturedAppRightMetadata();
	if (!featuredAppRightMetadata) {
		console.error("Unable to resolve Featured App Right metadata required for unlocking amulets.");
		return null;
	}

	const results = [];
	const endpoint = `${config.VALIDATOR_API_URL}/v2/commands/submit-and-wait`;

	// Unlock each locked amulet
	for (const holdingData of lockedHoldings) {
		const createdEvent = holdingData.contractEntry?.JsActiveContract?.createdEvent;
		if (!createdEvent) {
			console.warn(`Skipping invalid locked amulet holding data`);
			continue;
		}

		const lockedAmuletCid = createdEvent.contractId;
		const lockedAmuletTemplateId = createdEvent.templateId || null;
		const quantity = parseFloat(createdEvent.createArgument?.amount?.initialAmount || createdEvent.createArgument?.amulet?.amount?.initialAmount || 0);

		// Unlock this locked amulet

		// Build full template IDs using actual values from contracts, with env fallback
		const amuletRulesTemplateId = amuletRulesData.amulet_rules.contract.template_id || "#splice-amulet:Splice.AmuletRules:AmuletRules";
		const openMiningRoundTemplateId = miningRound.templateId || "#splice-amulet:Splice.Round:OpenMiningRound";
		// FeaturedAppRight template ID - NO FALLBACK, must come from env
		const featuredAppRightTemplateId = config.TEMPLATE_ID_FEATURED_APP;
		const lockedTemplateId = lockedAmuletTemplateId || "#splice-amulet:Splice.Amulet:LockedAmulet";

		const schema = getSchema("unlockAmulet");
		const templateVars = {
			locked_amulet_cid: lockedAmuletCid,
			locked_amulet_template_id: lockedTemplateId,
			open_mining_round_cid: miningRound.contractId,
			investor: party,
			user_id: config.AUTH0_USER_ID,
			randomUUID: randomUUID(),
			// Template IDs for disclosed contracts
			amulet_rules_template_id: amuletRulesTemplateId,
			open_mining_round_template_id: openMiningRoundTemplateId,
			featured_app_right_template_id: featuredAppRightTemplateId,
			// Amulet Rules disclosure
			transfer_rules_cid: amuletRulesData.amulet_rules.contract.contract_id,
			transfer_rules_disclosure: amuletRulesData.amulet_rules.contract.created_event_blob,
			transfer_rules_synchronizer: amuletRulesData.amulet_rules.domain_id,
			// Open Mining Round disclosure
			open_mining_round_disclosure: miningRound.createdEventBlob,
			open_mining_round_synchronizer: miningRound.synchronizerId,
			// Featured App Right disclosure
			featured_app_right_cid: featuredAppRightMetadata.contractCid,
			featured_app_right_disclosure: featuredAppRightMetadata.disclosureCid,
			featured_app_right_synchronizer: featuredAppRightMetadata.synchronizerId,
		};

		const command = replaceTemplateVariables(schema, templateVars);

		if (returnCommand) {
			results.push({ command, endpoint, contractId: lockedAmuletCid, amount: quantity });
			continue;
		}

		const headers = await buildHeaders();

		try {
			const response = await axios.post(endpoint, command, { headers });
			// Unlock successful
			results.push({ success: true, contractId: lockedAmuletCid, amount: quantity, data: response.data });
		} catch (error) {
			console.error(`Error unlocking amulet ${lockedAmuletCid}: ${error}`);
			if (error.response) {
				console.error(`Response data: ${JSON.stringify(error.response.data)}`);
			}
			results.push({ success: false, contractId: lockedAmuletCid, amount: quantity, error: error.message });
		}
	}

	if (returnCommand) {
		return results;
	}

	return results;
}

/**
 * Prepare holding contract IDs and formatted opts for depositFunds.
 * Selects the largest unlocked UTXOs until the requested amount is covered.
 *
 * @param {number|string} amount - The amount needed.
 * @param {string} symbol - Instrument symbol: "CC", "Amulet", "USDCx", "CBTC", etc.
 * @param {Object|null} [provider=null] - Optional wallet provider.
 * @returns {Promise<Object>} Object matching depositFunds opts shape, or { error }.
 */
export async function prepareDepositHoldings(amount, symbol) {
	const provider = resolveProvider(null);
	if (!provider && !config.VALIDATOR_SCAN_API_URL) {
		return { error: "prepareDepositHoldings: wallet provider is required to prepare holdings. Connect a wallet adapter." };
	}

	const party = getAdapterPartyId() ?? config.VALIDATOR_USER_PARTY_ID;
	if (!party) {
		return { error: "prepareDepositHoldings: party ID is required. Connect a wallet adapter or configure VALIDATOR_USER_PARTY_ID." };
	}

	// Normalize symbol: CC → Amulet
	let assetId = symbol?.trim();
	if (assetId === "CC") assetId = "Amulet";

	if (!instrumentCatalog[assetId]) {
		return { error: `prepareDepositHoldings: unsupported symbol "${symbol}"` };
	}

	const requiredAmount = parseFloat(amount);
	if (isNaN(requiredAmount) || requiredAmount <= 0) {
		return { error: "prepareDepositHoldings: amount must be a positive number" };
	}

	const isAmulet = assetId === "Amulet";
	const holdings = isAmulet ? await getAmuletHoldingsForParty(party, false, provider) : await getUtilityHoldingsForParty(party, false, provider);

	if (!holdings || holdings.length === 0) {
		return { error: `prepareDepositHoldings: no holdings found for party ${party}` };
	}

	// Parse holdings into { contractId, quantity } and filter out locked
	const parsed = [];
	for (const h of holdings) {
		const createdEvent = h.contractEntry?.JsActiveContract?.createdEvent;
		if (!createdEvent) continue;

		// Skip asset types that don't match the requested symbol
		const createdAssetId = createdEvent.createArgument?.instrument?.id || (isAmulet ? "Amulet" : null);
		if (createdAssetId !== assetId) continue;

		// Skip locked utility holdings
		if (!isAmulet && createdEvent.createArgument?.lock) continue;

		const quantity = isAmulet ? parseFloat(createdEvent.createArgument?.amount?.initialAmount || 0) : parseFloat(createdEvent.createArgument?.amount || 0);

		if (quantity <= 0) continue;

		parsed.push({
			contractId: normalizeContractId(createdEvent.contractId),
			quantity,
		});
	}

	if (parsed.length === 0) {
		return { error: `prepareDepositHoldings: no unlocked holdings found for ${assetId}` };
	}

	// Sort descending — largest first
	parsed.sort((a, b) => b.quantity - a.quantity);

	// Greedily select until we cover the required amount
	const selected = [];
	let cumulative = 0;
	for (const holding of parsed) {
		selected.push(holding.contractId);
		cumulative += holding.quantity;
		if (cumulative >= requiredAmount) break;
	}

	if (cumulative < requiredAmount) {
		return {
			error: `prepareDepositHoldings: insufficient balance. Have ${cumulative}, need ${requiredAmount}`,
			data: { totalBalance: cumulative, requiredAmount },
		};
	}

	const formattedAmount = parseFloat(parseFloat(amount).toFixed(10)).toString();

	return {
		sender: party,
		receiver: party,
		assetId,
		amount: formattedAmount,
		holdingCids: selected,
	};
}

/**
 * Create a CIP-56 Allocation.
 * Fetches the AllocationFactory, then exercises AllocationFactory_Allocate
 * to lock holdings for a settlement leg.
 *
 * For Amulet: uses the Scan API (VALIDATOR_SCAN_API_URL) with DSO as admin.
 * For utility tokens: uses the Registry API from the instrument catalog with registrar as admin.
 *
 * Flow:
 *   1. Resolve admin and factory URL (Scan API for Amulet, Registry API for utility).
 *   2. POST .../allocation-instruction/v1/allocation-factory to get
 *      factoryId, choiceContextData, and disclosedContracts.
 *   3. Build an ExerciseCommand on the AllocationFactory interface with the
 *      AllocationFactory_Allocate choice.
 *   4. Submit via the Ledger API submit-and-wait endpoint.
 *
 * @param {Object} opts
 * @param {string} opts.sender       - The party allocating the assets.
 * @param {string} opts.receiver     - The counterparty receiving in the settlement leg.
 * @param {string} opts.assetId      - Instrument symbol (e.g. "USDCx", "Amulet").
 * @param {string|number} opts.amount - Amount to allocate.
 * @param {string[]} opts.holdingCids - Holding contract IDs that fund the allocation.
 * @param {string} opts.settlementId - A reference id for the settlement.
 * @param {string} opts.transferLegId - Unique id for this transfer leg.
 * @param {string} [opts.allocateBefore] - ISO timestamp; deadline for allocation (default: +1 h).
 * @param {string} [opts.settleBefore]   - ISO timestamp; deadline for settlement (default: +2 h).
 * @param {Object} [opts.disclosures]    - Pre-fetched Amulet disclosures (from /api/amulet/disclosures). Required for Amulet when no Scan API or ledger access.
 * @param {string} [opts.userId]         - Temple user ID. Falls back to wallet adapter → AUTH0_USER_ID → "temple".
 * @param {boolean} [returnCommand=false] - If true, returns { command, endpoint } instead of executing.
 * @returns {Promise<Object|null>} Ledger API response (or { error } on failure).
 */
export async function depositFunds(opts, returnCommand = false) {
	const { sender, receiver, assetId, amount, holdingCids = [], settlementId, transferLegId, allocateBefore, settleBefore, disclosures, userId } = opts;

	if (!sender || !assetId || !amount) {
		const msg = "depositFunds: sender, assetId, and amount are required";
		console.error(msg);
		return { error: msg };
	}

	const isAmulet = assetId === "Amulet";

	// --- 1. Resolve admin and factory ---
	// Amulet: admin = DSO, factory from Scan API (or ledger on localhost)
	// Utility: admin = registrar from instrument catalog, factory from Registry API (or ledger on localhost)
	let admin;
	let factoryContractId;
	let choiceContextData = {};
	let disclosedContracts = [];

	if (isAmulet) {
		admin = config.VALIDATOR_DSO_PARTY_ID;
	} else {
		const instrumentDef = resolveInstrumentDefinition(assetId);
		const networkContracts = instrumentDef?.[config.NETWORK];
		admin = networkContracts?.registrar || getInstrumentRegistrar(assetId) || config.VALIDATOR_REGISTRAR_PARTY_ID;
	}

	// --- 2. Build the choice arguments upfront ---
	// Needed both for the factory request (Registry API requires the full args) and the exercise command
	const now = new Date();
	const allocateDeadline = allocateBefore || new Date(now.getTime() + 60 * 60 * 1000).toISOString();
	const settleDeadline = settleBefore || new Date(now.getTime() + 2 * 60 * 60 * 1000).toISOString();
	const executor = config.TEMPLE_PARTY_ID;
	const resolvedTransferLegId = transferLegId || randomUUID();
	const resolvedSettlementId = settlementId || `SR-${randomUUID()}`;

	const choiceArgument = {
		expectedAdmin: admin,
		allocation: {
			settlement: {
				executor: executor,
				settlementRef: {
					id: resolvedSettlementId,
					cid: null,
				},
				requestedAt: now.toISOString(),
				allocateBefore: allocateDeadline,
				settleBefore: settleDeadline,
				meta: { values: config.ALLOCATION_META_TAG ? { tag: config.ALLOCATION_META_TAG } : {} },
			},
			transferLegId: resolvedTransferLegId,
			transferLeg: {
				sender: sender,
				receiver: receiver || sender,
				amount: parseFloat(parseFloat(amount).toFixed(10)).toString(),
				instrumentId: {
					admin: admin,
					id: assetId,
				},
				meta: { values: {} },
			},
		},
		requestedAt: now.toISOString(),
		inputHoldingCids: holdingCids.map(normalizeContractId),
		extraArgs: {
			context: { values: {} },
			meta: { values: {} },
		},
	};

	// --- 3. Fetch AllocationFactory ---
	if (shouldUseLedgerForMetadata()) {
		// Localhost: resolve factory and context from ledger directly
		if (isAmulet) {
			const amuletCtx = await resolveAmuletContext({ investor: sender, holdingIds: holdingCids, transferAmount: Number(amount) });
			if (!amuletCtx) {
				const msg = "depositFunds: failed to resolve Amulet context from ledger";
				console.error(msg);
				return { error: msg };
			}
			// ExternalPartyAmuletRules is the AllocationFactory for Amulet
			factoryContractId = normalizeContractId(amuletCtx.externalAmuletRules.contractCid);
			choiceContextData = {
				values: {
					[amuletCtx.contextKeys.amuletRules]: { tag: "AV_ContractId", value: amuletCtx.amuletRules.contractCid },
					[amuletCtx.contextKeys.openRound]: { tag: "AV_ContractId", value: amuletCtx.openMiningRound.contractCid },
					[amuletCtx.contextKeys.featuredAppRight]: { tag: "AV_ContractId", value: amuletCtx.featuredAppRight.contractCid },
				},
			};
			disclosedContracts = [
				{
					templateId: amuletCtx.amuletRules.templateId,
					contractId: amuletCtx.amuletRules.contractCid,
					createdEventBlob: amuletCtx.amuletRules.disclosureCid,
					synchronizerId: amuletCtx.amuletRules.synchronizerId,
				},
				{
					templateId: amuletCtx.openMiningRound.templateId,
					contractId: amuletCtx.openMiningRound.contractCid,
					createdEventBlob: amuletCtx.openMiningRound.disclosureCid,
					synchronizerId: amuletCtx.openMiningRound.synchronizerId,
				},
				{
					templateId: amuletCtx.externalAmuletRules.templateId,
					contractId: amuletCtx.externalAmuletRules.contractCid,
					createdEventBlob: amuletCtx.externalAmuletRules.disclosureCid,
					synchronizerId: amuletCtx.externalAmuletRules.synchronizerId,
				},
			];
			if (amuletCtx.featuredAppRight?.contractCid && amuletCtx.featuredAppRight?.disclosureCid) {
				disclosedContracts.push({
					templateId: amuletCtx.featuredAppRight.templateId,
					contractId: amuletCtx.featuredAppRight.contractCid,
					createdEventBlob: amuletCtx.featuredAppRight.disclosureCid,
					synchronizerId: amuletCtx.featuredAppRight.synchronizerId,
				});
			}
		} else {
			// Utility token: resolve from ledger
			const instrumentDef = resolveInstrumentDefinition(assetId);
			const registrar = instrumentDef?.[config.NETWORK]?.registrar || getInstrumentRegistrar(assetId) || config.VALIDATOR_REGISTRAR_PARTY_ID;
			const [allocFactory, instConfig] = await Promise.all([resolveUtilityAllocationFactory(registrar, sender), resolveUtilityInstrumentConfiguration(assetId, registrar)]);
			if (!allocFactory) {
				const msg = `depositFunds: no AllocationFactory found on ledger for ${assetId}`;
				console.error(msg);
				return { error: msg };
			}
			factoryContractId = normalizeContractId(allocFactory.contractCid);
			if (instConfig) {
				choiceContextData = {
					values: {
						[DEFAULT_UTILITY_CONTEXT_KEYS.instrumentConfiguration]: { tag: "AV_ContractId", value: instConfig.contractCid },
						[DEFAULT_UTILITY_CONTEXT_KEYS.instrumentConfigurationPrefixed]: { tag: "AV_ContractId", value: instConfig.contractCid },
						[DEFAULT_UTILITY_CONTEXT_KEYS.senderCredentials]: { tag: "AV_List", value: [] },
						[DEFAULT_UTILITY_CONTEXT_KEYS.senderCredentialsPrefixed]: { tag: "AV_List", value: [] },
						[DEFAULT_UTILITY_CONTEXT_KEYS.receiverCredentials]: { tag: "AV_List", value: [] },
						[DEFAULT_UTILITY_CONTEXT_KEYS.receiverCredentialsPrefixed]: { tag: "AV_List", value: [] },
					},
				};
				disclosedContracts.push({ templateId: null, contractId: instConfig.contractCid, createdEventBlob: instConfig.disclosureCid, synchronizerId: instConfig.synchronizerId });
			}
			disclosedContracts.push({ templateId: null, contractId: allocFactory.contractCid, createdEventBlob: allocFactory.disclosureCid, synchronizerId: allocFactory.synchronizerId });
		}
	} else if (isAmulet && !config.VALIDATOR_SCAN_API_URL) {
		// Amulet via disclosures (FE/proxy path — no API access)
		// Use pre-fetched disclosures if provided, otherwise fetch from API
		const factoryData = disclosures?.disclosures || disclosures || null;
		let resolvedFactoryData = factoryData;

		if (!resolvedFactoryData?.factoryId) {
			const disclosuresResult = await getDisclosures(sender);
			resolvedFactoryData = disclosuresResult?.disclosures;
			if (disclosuresResult?.error || !resolvedFactoryData?.factoryId) {
				const detail = disclosuresResult?.message || disclosuresResult?.error || "missing factoryId in response";
				const msg = `depositFunds: failed to resolve Amulet disclosures. Pass disclosures directly or ensure API access: ${detail}`;
				console.error(msg);
				return { error: msg };
			}
		}

		factoryContractId = normalizeContractId(resolvedFactoryData.factoryId);
		choiceContextData = resolvedFactoryData.choiceContext?.choiceContextData || {};
		disclosedContracts = (resolvedFactoryData.choiceContext?.disclosedContracts || []).map((dc) => ({
			templateId: dc.templateId,
			contractId: dc.contractId,
			createdEventBlob: dc.createdEventBlob,
			synchronizerId: dc.synchronizerId,
		}));
	} else {
		// Remote: use Scan API (Amulet) or Registry API (utility)
		let factoryUrl;
		let factoryHeaders = {};

		if (isAmulet) {
			factoryUrl = `${config.VALIDATOR_SCAN_API_URL}/registry/allocation-instruction/v1/allocation-factory`;
			factoryHeaders = await buildHeaders();
		} else {
			const instrumentDef = resolveInstrumentDefinition(assetId);
			const networkContracts = instrumentDef?.[config.NETWORK];
			const registryAPI = networkContracts?.registryAPI;

			if (!registryAPI) {
				const msg = `depositFunds: no registryAPI defined for ${assetId} on ${config.NETWORK}`;
				console.error(msg);
				return { error: msg };
			}
			factoryUrl = `${registryAPI}/allocation-instruction/v1/allocation-factory`;
		}

		const factoryBody = {
			choiceArguments: isAmulet ? {} : choiceArgument,
			excludeDebugFields: true,
		};
		let factoryData;
		try {
			const factoryResponse = await axios.post(factoryUrl, factoryBody, Object.keys(factoryHeaders).length > 0 ? { headers: factoryHeaders } : undefined);
			factoryData = factoryResponse.data;
		} catch (error) {
			const detail = error.response?.data ? JSON.stringify(error.response.data) : error.message;
			const msg = `depositFunds: error fetching allocation factory from ${isAmulet ? "Scan API" : "Registry API"}: ${detail}`;
			console.error(msg);
			return { error: msg };
		}

		factoryContractId = normalizeContractId(factoryData?.factoryId);
		choiceContextData = factoryData?.choiceContext?.choiceContextData || {};
		disclosedContracts = (factoryData?.choiceContext?.disclosedContracts || []).map((dc) => ({
			templateId: dc.templateId,
			contractId: dc.contractId,
			createdEventBlob: dc.createdEventBlob,
			synchronizerId: dc.synchronizerId,
		}));
	}

	if (!factoryContractId) {
		const msg = "depositFunds: allocation factory response missing factoryId";
		console.error(msg);
		return { error: msg };
	}

	// --- 4. Build the ExerciseCommand for AllocationFactory_Allocate ---
	// Merge the resolved choiceContextData into the extraArgs
	choiceArgument.extraArgs.context = choiceContextData;

	const command = {
		commands: [
			{
				ExerciseCommand: {
					templateId: "#splice-api-token-allocation-instruction-v1:Splice.Api.Token.AllocationInstructionV1:AllocationFactory",
					contractId: factoryContractId,
					choice: "AllocationFactory_Allocate",
					choiceArgument: choiceArgument,
				},
			},
		],
		commandId: randomUUID(),
		userId: userId || getUserId() || config.AUTH0_USER_ID || "temple",
		applicationId: "temple",
		actAs: [sender],
		disclosedContracts: disclosedContracts,
	};

	dedupeDisclosedContracts(command);
	const endpoint = `${config.VALIDATOR_API_URL}/v2/commands/submit-and-wait`;

	if (returnCommand) {
		return { command, endpoint };
	}

	// Auto-submit via wallet adapter if available
	if (getWalletAdapter()) {
		try {
			return await submitCommand(command);
		} catch (error) {
			const msg = `depositFunds: wallet adapter submission failed: ${error.message}`;
			console.error(msg);
			return { error: msg };
		}
	}

	const headers = await buildHeaders();
	try {
		const response = await axios.post(endpoint, command, { headers });
		return response.data;
	} catch (error) {
		const errorData = error?.response?.data;
		const errorDetail = errorData ? JSON.stringify(errorData, null, 2) : error.message;
		const msg = `depositFunds: error submitting command: ${errorDetail}`;
		console.error(msg);
		return { error: msg };
	}
}

// Withdrawal functions moved to src/canton/withdrawals.ts
export { finalizeWithdrawFunds, withdrawFunds, withdrawDelegation } from "../../dist/canton/withdrawals.js";

// ─── Onboarding & Delegation ─────────────────────────────────────────────────

/**
 * Create an OnboardingRequest contract on the ledger.
 *
 * The user creates this contract to request delegation to the operator.
 * The operator then exercises AcceptOnboarding to produce a Delegation contract.
 *
 * Template: Temple.Settlement.Onboarding:OnboardingRequest
 *
 * @param {object} user     - User info: { partyId, userId }. partyId is the Daml party; userId is for command submission (e.g. Auth0 sub).
 * @param {string} user.partyId - The Daml party representing the user (controller of the onboarding request).
 * @param {string} user.userId  - The user ID for command submission (e.g. Auth0 sub). Defaults to config.AUTH0_USER_ID or "temple".
 * @param {boolean} [returnCommand=false] - If true, returns { command, endpoint } without submitting.
 * @returns {Promise<Object>} Ledger API response, or { error } on failure.
 */
export async function onboardUser(user = {}, returnCommand = false) {
	user.partyId = user.partyId ?? getAdapterPartyId() ?? config.VALIDATOR_USER_PARTY_ID;
	user.userId = user.userId ?? getUserId() ?? config.AUTH0_USER_ID ?? "temple";
	if (!user.partyId) {
		const msg = "onboardUser: user party is required. Pass it directly, set WALLET_ADAPTER, or configure VALIDATOR_USER_PARTY_ID.";
		console.error(msg);
		return { error: msg };
	}

	const operator = config.TEMPLE_PARTY_ID;
	if (!operator) {
		const msg = "onboardUser: TEMPLE_PARTY_ID (operator) is not configured";
		console.error(msg);
		return { error: msg };
	}

	const command = {
		commands: [
			{
				CreateCommand: {
					templateId: `${config.PACKAGE_NAME_SETTLEMENT_IMPL}:Temple.Settlement.Onboarding:OnboardingRequest`,
					createArguments: {
						user: user.partyId,
						operator: operator,
					},
				},
			},
		],
		commandId: randomUUID(),
		userId: user.userId,
		applicationId: "temple",
		actAs: [user.partyId],
	};

	const endpoint = `${config.VALIDATOR_API_URL}/v2/commands/submit-and-wait`;

	if (returnCommand) {
		return { command, endpoint };
	}

	// Auto-submit via wallet adapter if available
	let submitResult;
	if (getWalletAdapter()) {
		try {
			submitResult = await submitCommand(command);
		} catch (error) {
			const msg = `onboardUser: wallet adapter submission failed: ${error.message}`;
			console.error(msg);
			return { error: msg };
		}
	} else {
		const headers = await buildHeaders();
		try {
			const response = await axios.post(endpoint, command, { headers });
			submitResult = response.data;
		} catch (error) {
			const errorData = error?.response?.data;
			const errorDetail = errorData ? JSON.stringify(errorData, null, 2) : error.message;
			const msg = `onboardUser: error submitting command: ${errorDetail}`;
			console.error(msg);
			return { error: msg };
		}
	}

	// Poll isUserOnboarded every 5s for up to 60s
	const pollInterval = 5_000;
	const maxAttempts = 12;
	for (let attempt = 1; attempt <= maxAttempts; attempt++) {
		await new Promise((resolve) => setTimeout(resolve, pollInterval));
		const delegation = await isUserOnboarded(user.partyId);
		if (delegation && !delegation.error) {
			console.log(`onboardUser: onboarding confirmed after ${attempt * 5}s`);
			return { ...submitResult, delegation };
		}
		if (attempt < maxAttempts) {
			console.log(`onboardUser: waiting for onboarding confirmation (${attempt * 5}s)...`);
		}
	}

	console.warn("onboardUser: command submitted but onboarding not confirmed within 60s");
	return { ...submitResult, warning: "Onboarding request submitted but delegation not confirmed within 60s" };
}


/**
 * Check if a user is already onboarded (has an active Delegation contract with the operator).
 *
 * Queries active contracts for a Delegation template where the user is a signatory.
 * Returns the Delegation contract if found, or null if not onboarded.
 *
 * @param {string} user - The party to check.
 * @param {boolean} [returnCommand=false] - If true, returns { command, endpoint } without executing.
 * @returns {Promise<Object|null>} The Delegation contract data if onboarded, null if not, or { error } on failure.
 */
export async function isUserOnboarded(user = null, returnCommand = false) {
	// ── API path: use delegation endpoint when no custom validator ──
	if (!config.VALIDATOR_API_URL || getWalletAdapter()) {
		const result = await getDelegation();
		if (result?.error) return null;
		return result?.delegation || null;
	}

	// ── Custom validator path: query ledger for delegation contract ──
	if (!user) {
		const msg = "isUserOnboarded: user party is required";
		console.error(msg);
		return { error: msg };
	}

	const ledgerEnd = await getLedgerEnd();
	if (ledgerEnd === null) {
		const msg = "isUserOnboarded: cannot retrieve ledger end";
		console.error(msg);
		return { error: msg };
	}

	const templateId = `${config.PACKAGE_NAME_SETTLEMENT_IMPL}:Temple.Settlement.Delegation:Delegation`;

	const command = {
		verbose: true,
		activeAtOffset: ledgerEnd,
		filter: {
			filtersByParty: {
				[user]: {
					cumulative: [
						{
							identifierFilter: {
								TemplateFilter: {
									value: {
										templateId,
										includeCreatedEventBlob: false,
									},
								},
							},
						},
					],
				},
			},
		},
	};

	const endpoint = `${config.VALIDATOR_API_URL}/v2/state/active-contracts`;

	if (returnCommand) {
		return { command, endpoint };
	}

	const headers = await buildHeaders();
	try {
		const response = await axios.post(endpoint, command, { headers });
		const contracts = unwrapActiveContracts(response.data);

		// Filter to find delegation with our operator
		const operator = config.TEMPLE_PARTY_ID;
		const delegation = contracts.find((entry) => {
			const payload = entry?.contractEntry?.JsActiveContract?.createdEvent?.createArgument;
			return payload?.user === user && payload?.operator === operator;
		});

		return delegation || null;
	} catch (error) {
		const errorData = error?.response?.data;
		const errorDetail = errorData ? JSON.stringify(errorData, null, 2) : error.message;
		const msg = `isUserOnboarded: error querying active contracts: ${errorDetail}`;
		console.error(msg);
		return { error: msg };
	}
}

/**
 * Get all balances for a party, grouped by asset type.
 * Returns Amulet, locked Amulet, and utility token holdings with totals.
 * @param {string|null} [party=null] - The party ID. Optional if authenticated via REST API or wallet adapter is set.
 * @param {Object|null} [provider=null] - Optional wallet provider. Falls back to wallet adapter if set.
 * @returns {Promise<Object>} Balances grouped by registrar and asset, each with total_balance, holdings, etc.
 */
export async function getUserBalances(party = null, provider = null) {
	provider = resolveProvider(provider);
	if (!provider && !config.VALIDATOR_API_URL) {
		const msg = "getUserBalances: no provider or API URL configured. Set up a wallet adapter or configure VALIDATOR_API_URL.";
		console.error(msg);
		return { error: msg };
	}
	party = party ?? getAdapterPartyId() ?? config.VALIDATOR_USER_PARTY_ID;
	if (!party) {
		return { error: "Party ID is required. Pass it directly, connect a wallet adapter, or configure VALIDATOR_USER_PARTY_ID." };
	}
	// Get all supported instrument IDs from the catalog
	const supportedInstruments = new Set(Object.keys(instrumentCatalog));

	// ── Provider path: use getHolding() for all balances ──
	if (provider) {
		const holdings = await provider.getHolding();

		const balances = [];
		for (const h of holdings || []) {
			const instrumentId = h.instrument_id?.id;
			const symbol = h.symbol;

			// Only include instruments in our catalog
			const catalogKey = supportedInstruments.has(instrumentId) ? instrumentId : instrumentIdToSymbol[instrumentId] || null;
			if (!catalogKey) continue;

			const catalogEntry = instrumentCatalog[catalogKey];
			const unlocked = parseFloat(h.total_unlocked_coin || "0");
			const locked = parseFloat(h.total_locked_coin || "0");

			balances.push({
				asset: catalogKey,
				symbol: symbol,
				total_balance: unlocked + locked,
				available_balance: unlocked,
				locked_balance: locked,
				dso: catalogEntry?.kind === "amulet" ? h.instrument_id?.admin || null : null,
				registrar: h.instrument_id?.admin || null,
				operator: catalogEntry?.operator || null,
				provider: catalogEntry?.provider || null,
				merge_warning: (h.total_utxos || 0) > 1,
				holdings: [],
				locked_holdings: [],
				utilityContext: null,
			});
		}

		return balances;
	}

	// ── Custom validator path: query individual contract types from ledger ──
	const promises = [getAmuletHoldingsForParty(party, false, null), getLockedAmuletHoldingsForParty(party, false, null), getUtilityHoldingsForParty(party, false, null)];

	const [amuletHoldings, lockedAmuletHoldings, utilityHoldings] = await Promise.all(promises);

	// Structure: assetsByKey[registrar][asset] = { ... }
	const assetsByKey = {};

	// Helper to initialize asset entry
	const initAssetEntry = (key, asset, metadata = {}) => {
		if (!assetsByKey[key]) {
			assetsByKey[key] = {};
		}
		if (!assetsByKey[key][asset]) {
			assetsByKey[key][asset] = {
				available_balance: 0,
				locked_balance: 0,
				holdings: [],
				locked_holdings: [],
				registrar: metadata.registrar || null,
				dso: metadata.dso || null,
				operator: metadata.operator || null,
				provider: metadata.provider || null,
				utilityContext: null,
			};
		}
		return assetsByKey[key][asset];
	};

	// Process unlocked Amulet holdings
	for (const holdingData of amuletHoldings || []) {
		const createdEvent = holdingData.contractEntry?.JsActiveContract?.createdEvent;
		if (!createdEvent) continue;

		const quantity = parseFloat(createdEvent.createArgument?.amount?.initialAmount || createdEvent.createArgument?.amulet?.amount?.initialAmount || 0);
		const dso = createdEvent.createArgument?.dso || createdEvent.createArgument?.amulet?.dso || "Unknown DSO";
		const synchronizerId = holdingData.contractEntry?.JsActiveContract?.synchronizerId || null;
		const disclosure = createdEvent.createdEventBlob || null;

		const catalogEntry = instrumentCatalog["Amulet"];
		const key = dso; // For Amulet, use DSO as the key
		const entry = initAssetEntry(key, "Amulet", {
			registrar: catalogEntry?.registrar || dso,
			dso: dso,
			operator: catalogEntry?.operator || null,
			provider: catalogEntry?.provider || null,
		});

		entry.available_balance += quantity;
		entry.holdings.push({
			contractId: createdEvent.contractId,
			quantity: quantity,
			disclosure,
			synchronizer: synchronizerId,
			dso: dso,
		});
	}

	// Process locked Amulet holdings
	for (const holdingData of lockedAmuletHoldings || []) {
		const createdEvent = holdingData.contractEntry?.JsActiveContract?.createdEvent;
		if (!createdEvent) continue;

		const quantity = parseFloat(createdEvent.createArgument?.amulet?.amount?.initialAmount || 0);
		const dso = createdEvent.createArgument?.amulet?.dso || "Unknown DSO";
		const synchronizerId = holdingData.contractEntry?.JsActiveContract?.synchronizerId || null;
		const disclosure = createdEvent.createdEventBlob || null;
		const lockHolder = createdEvent.createArgument?.lock?.holders?.[0] || null;
		const lockExpiry = createdEvent.createArgument?.lock?.expiresAt || null;

		const catalogEntry = instrumentCatalog["Amulet"];
		const key = dso;
		const entry = initAssetEntry(key, "Amulet", {
			registrar: catalogEntry?.registrar || dso,
			dso: dso,
			operator: catalogEntry?.operator || null,
			provider: catalogEntry?.provider || null,
		});

		entry.locked_balance += quantity;
		entry.locked_holdings.push({
			contractId: createdEvent.contractId,
			quantity: quantity,
			disclosure,
			synchronizer: synchronizerId,
			dso: dso,
			lockHolder,
			lockExpiry,
		});
	}

	// Process utility holdings - check lock status to separate locked/unlocked
	for (const holdingData of utilityHoldings || []) {
		const createdEvent = holdingData.contractEntry?.JsActiveContract?.createdEvent;
		if (!createdEvent) continue;

		const quantity = parseFloat(createdEvent.createArgument?.amount || 0);
		let asset = createdEvent.createArgument?.instrument?.id || "Unknown Utility";
		// Resolve on-chain instrument ID to catalog symbol (e.g. UUID -> "SBC")
		asset = instrumentIdToSymbol[asset] || asset;
		// Normalize "Amulet" utility tokens to "tAmulet" to distinguish from native Amulet
		if (asset === "Amulet") {
			asset = "tAmulet";
		}

		// Only include supported instruments
		if (!supportedInstruments.has(asset)) continue;

		const registrar = createdEvent.createArgument?.instrument?.source || "Unknown Registrar";
		const synchronizerId = holdingData.contractEntry?.JsActiveContract?.synchronizerId || null;
		const disclosure = createdEvent.createdEventBlob || null;

		// Check if the holding is locked
		const lockData = createdEvent.createArgument?.lock || null;
		const isLocked = lockData !== null;
		const lockHolder = lockData?.holders?.[0] || null;
		const lockExpiry = lockData?.expiresAt || null;

		const catalogEntry = instrumentCatalog[asset];
		const key = registrar;
		const entry = initAssetEntry(key, asset, {
			registrar: catalogEntry?.registrar || registrar,
			dso: null, // Utility tokens don't have DSO
			operator: catalogEntry?.operator || null,
			provider: catalogEntry?.provider || null,
		});

		if (isLocked) {
			entry.locked_balance += quantity;
			entry.locked_holdings.push({
				contractId: createdEvent.contractId,
				quantity: quantity,
				disclosure,
				synchronizer: synchronizerId,
				registrar: registrar,
				lockHolder,
				lockExpiry,
			});
		} else {
			entry.available_balance += quantity;
			entry.holdings.push({
				contractId: createdEvent.contractId,
				quantity: quantity,
				disclosure,
				synchronizer: synchronizerId,
				registrar: registrar,
			});
		}
	}

	// Resolve utility contexts (requires ledger API access — provider path returns early above)
	const utilityContextPromises = new Map();
	for (const key of Object.keys(assetsByKey)) {
		for (const asset of Object.keys(assetsByKey[key])) {
			if (asset === "Amulet") {
				continue;
			}
			const assetEntry = assetsByKey[key][asset];
			if (assetEntry.utilityContext) {
				continue;
			}
			const registrar = assetEntry.registrar || key;
			const cacheKey = `${registrar || "*"}::${asset}`;
			if (!utilityContextPromises.has(cacheKey)) {
				utilityContextPromises.set(
					cacheKey,
					resolveUtilityCip56Context({
						party,
						registrar,
						utilityAsset: asset,
					}),
				);
			}
		}
	}

	// Await all utility context promises
	for (const [key, promise] of utilityContextPromises.entries()) {
		utilityContextPromises.set(key, await promise);
	}

	// Assign utility contexts
	for (const key of Object.keys(assetsByKey)) {
		for (const asset of Object.keys(assetsByKey[key])) {
			if (asset === "Amulet") {
				continue;
			}
			const assetEntry = assetsByKey[key][asset];
			if (assetEntry.utilityContext) {
				continue;
			}
			const registrar = assetEntry.registrar || key;
			const cacheKey = `${registrar || "*"}::${asset}`;
			assetEntry.utilityContext = utilityContextPromises.get(cacheKey) || null;
			if (!assetEntry.registrar && assetEntry.utilityContext?.registrar) {
				assetEntry.registrar = assetEntry.utilityContext.registrar;
			}
		}
	}

	// Build the final balances array
	const balances = [];
	for (const key in assetsByKey) {
		for (const asset in assetsByKey[key]) {
			const assetEntry = assetsByKey[key][asset];
			balances.push({
				asset: asset,
				total_balance: assetEntry.available_balance + assetEntry.locked_balance,
				available_balance: assetEntry.available_balance,
				locked_balance: assetEntry.locked_balance,
				dso: assetEntry.dso,
				registrar: assetEntry.registrar,
				operator: assetEntry.operator,
				provider: assetEntry.provider,
				merge_warning: assetEntry.holdings.length > 1,
				holdings: assetEntry.holdings,
				locked_holdings: assetEntry.locked_holdings,
				utilityContext: assetEntry.utilityContext,
			});
		}
	}

	return balances;
}
