// @temple-digital-group/temple-canton-js/src/canton/walletAdapter.js
/**
 * Wallet Adapter Module
 *
 * Provides a unified interface for wallet interactions (reads and writes).
 * Currently supports Loop Wallet. Designed to support additional wallets in the future.
 *
 * Server-side (Loop): uses executeTransaction() — signs locally with private key
 * Client-side (Loop): uses submitTransaction() — delegates signing to Loop Wallet via WebSocket
 *
 * Detection: Loop server-side has null connection.ticketId and connection.ws
 */

let _adapter = null;
let _isServer = false;

/**
 * Set the wallet adapter (e.g., Loop SDK instance).
 * Auto-detects server vs client mode.
 */
export function setWalletAdapter(adapter) {
	_adapter = adapter;
	_isServer = adapter ? detectServerSide(adapter) : false;
}

/**
 * Get the current wallet adapter.
 */
export function getWalletAdapter() {
	return _adapter;
}

/**
 * Returns the provider sub-object used for read operations (getActiveContracts).
 * For Loop, this is adapter.provider.
 */
export function getAdapterProvider() {
	if (!_adapter) return null;
	return _adapter.provider || _adapter;
}

/**
 * Check if running in server mode.
 */
export function isServerMode() {
	return _isServer;
}

/**
 * Detect if the adapter is server-side.
 * Loop-specific: server-side has null connection.ticketId and connection.ws
 */
function detectServerSide(adapter) {
	if (adapter?.connection) {
		return adapter.connection.ticketId == null && adapter.connection.ws == null;
	}
	// If no connection object, assume server-side (e.g., headless/API usage)
	return true;
}

/**
 * Get the party ID from the wallet adapter.
 * For Loop, this is on adapter.connection.partyId or adapter.partyId.
 */
export function getAdapterPartyId() {
	if (!_adapter) return null;
	return _adapter.provider?.party_id || _adapter.session?.partyId || null;
}

/**
 * Submit a command via the wallet adapter.
 * Server-side: calls adapter.executeTransaction(command)
 * Client-side: calls adapter.provider.submitTransaction(command)
 */
export async function submitCommand(command) {
	if (!_adapter) {
		throw new Error("[Temple SDK] No wallet adapter configured. Call initialize() with WALLET_ADAPTER or setWalletAdapter() first.");
	}

	if (_isServer && typeof _adapter.executeTransaction === 'function') {
		return _adapter.executeTransaction(command);
	}

	const provider = _adapter.provider || _adapter;
	if (typeof provider.submitTransaction === 'function') {
		return provider.submitTransaction(command);
	}

	throw new Error("[Temple SDK] Wallet adapter does not support transaction submission.");
}
